<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Introduction | Multi-modal Data Integration</title>
  <meta name="description" content="Covers patient classification using multi-modal data integration" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Introduction | Multi-modal Data Integration" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Covers patient classification using multi-modal data integration" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Introduction | Multi-modal Data Integration" />
  
  <meta name="twitter:description" content="Covers patient classification using multi-modal data integration" />
  

<meta name="author" content="Instructor: Shraddha Pai" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
<link rel="prev" href="index.html"/>
<link rel="next" href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"/>
<script src="libs/header-attrs-2.8/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">CBW CAN Multi-modal Data Integration</a></li>

<li class="divider"></li>
<li><a href="index.html#section"></a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i>Introduction</a>
<ul>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#course-materials"><i class="fa fa-check"></i>Course materials</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#data"><i class="fa fa-check"></i>Data</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#labs"><i class="fa fa-check"></i>Labs</a>
<ul>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#lab-1-a-gentle-introduction"><i class="fa fa-check"></i>LAB 1: A gentle introduction</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#lab-2-pathway-features-custom-similarity-metrics"><i class="fa fa-check"></i>LAB 2: Pathway features &amp; custom similarity metrics</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#instructions"><i class="fa fa-check"></i>Instructions</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"><a href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"><i class="fa fa-check"></i>Module 11 Lab 1: Three-way classifier from four genomic layers</a>
<ul>
<li class="chapter" data-level="" data-path="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"><a href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html#introduction-1"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"><a href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html#get-and-prepare-data"><i class="fa fa-check"></i>Get and prepare data</a></li>
<li class="chapter" data-level="" data-path="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"><a href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html#create-feature-design-rules"><i class="fa fa-check"></i>Create feature design rules</a>
<ul>
<li class="chapter" data-level="" data-path="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"><a href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html#grouplist"><i class="fa fa-check"></i>groupList</a></li>
<li class="chapter" data-level="" data-path="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"><a href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html#define-patient-similarity-for-each-network"><i class="fa fa-check"></i>Define patient similarity for each network</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"><a href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html#build-predictor"><i class="fa fa-check"></i>Build predictor</a></li>
<li class="chapter" data-level="" data-path="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"><a href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html#examine-results"><i class="fa fa-check"></i>Examine results</a>
<ul>
<li class="chapter" data-level="" data-path="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"><a href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html#get-model-results"><i class="fa fa-check"></i>Get model results</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"><a href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html#conclusion"><i class="fa fa-check"></i>Conclusion</a></li>
<li class="chapter" data-level="" data-path="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html"><a href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html#sessioninfo"><i class="fa fa-check"></i>sessionInfo</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-11-lab-2-pathway-level-features.html"><a href="module-11-lab-2-pathway-level-features.html"><i class="fa fa-check"></i>Module 11 Lab 2: Pathway-level features</a>
<ul>
<li class="chapter" data-level="" data-path="module-11-lab-2-pathway-level-features.html"><a href="module-11-lab-2-pathway-level-features.html#introduction-2"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-11-lab-2-pathway-level-features.html"><a href="module-11-lab-2-pathway-level-features.html#get-and-prepare-data-1"><i class="fa fa-check"></i>Get and prepare data</a></li>
<li class="chapter" data-level="" data-path="module-11-lab-2-pathway-level-features.html"><a href="module-11-lab-2-pathway-level-features.html#create-feature-design-rules-1"><i class="fa fa-check"></i>Create feature design rules</a>
<ul>
<li class="chapter" data-level="" data-path="module-11-lab-2-pathway-level-features.html"><a href="module-11-lab-2-pathway-level-features.html#rna-pathway-features"><i class="fa fa-check"></i>RNA: Pathway features</a></li>
<li class="chapter" data-level="" data-path="module-11-lab-2-pathway-level-features.html"><a href="module-11-lab-2-pathway-level-features.html#clinical-single-variables"><i class="fa fa-check"></i>Clinical: Single variables</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-11-lab-2-pathway-level-features.html"><a href="module-11-lab-2-pathway-level-features.html#build-predictor-1"><i class="fa fa-check"></i>Build predictor</a></li>
<li class="chapter" data-level="" data-path="module-11-lab-2-pathway-level-features.html"><a href="module-11-lab-2-pathway-level-features.html#examine-results-1"><i class="fa fa-check"></i>Examine results</a></li>
<li class="chapter" data-level="" data-path="module-11-lab-2-pathway-level-features.html"><a href="module-11-lab-2-pathway-level-features.html#visualize-top-pathways-in-cytoscape"><i class="fa fa-check"></i>Visualize top pathways in Cytoscape</a></li>
<li class="chapter" data-level="" data-path="module-11-lab-2-pathway-level-features.html"><a href="module-11-lab-2-pathway-level-features.html#sessioninfo-1"><i class="fa fa-check"></i>sessionInfo</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Multi-modal Data Integration</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This book contains lab exercises for Module 11: Multi-modal Data Integration, as part of the <a href="https://bioinformatics.ca/workshops/2021-cancer-analysis/">Canadian Bioinformatics Workshop Cancer Analysis (CAN) Workshop</a>.</p>
<div id="course-materials" class="section level2">
<h2>Course materials</h2>
<p>You can download course slides here <strong>LINK TBA</strong></p>
<p>We will have two lab exercises, where we build patient classifiers using multi-modal data integration, using netDx.<span class="citation"><sup><a href="#ref-Pai2019-iu" role="doc-biblioref">1</a>,<a href="#ref-Pai2021-um" role="doc-biblioref">2</a></sup></span></p>
</div>
<div id="data" class="section level2">
<h2>Data</h2>
<style>
p.caption {
  font-size: 0.6em;
}
</style>
<p>For both lab exercises, we will use multi-modal breast cancer data from The Cancer Genome Atlas,<span class="citation"><sup><a href="#ref-noauthor_2012-nc" role="doc-biblioref">3</a></sup></span> and classify tumours by molecular subtypes originally defined using gene expression profiles.</p>
<p>In particular, we will focus on classifying tumours as being of either <em>Luminal A</em>, <em>Luminal B</em>, or <em>Basal-like</em> subtypes. These subtypes vary by prevalence, prognosis for survival and relapse, activated cellular signaling pathways, immunohistochemical markers and treatment options.<span class="citation"><sup><a href="#ref-Yersal2014-km" role="doc-biblioref">4</a></sup></span></p>
<div class="figure"><span id="fig:unnamed-chunk-1"></span>
<img src="images/intrinsic-and-molecular-subtypes-of-breast-cancer.png" alt="Reprinted from [Intrinsic and Molecular Subtypes of Breast Cancer](https://app.biorender.com/biorender-templates/figures/5f15ea4438c5ef002876ab4b/t-5f872409fb2c3900a82e109e-intrinsic-and-molecular-subtypes-of-breast-cancer), by BioRender, October 2020, Copyright 2021 by BioRender." width="601" />
<p class="caption">
Figure 1: Reprinted from <a href="https://app.biorender.com/biorender-templates/figures/5f15ea4438c5ef002876ab4b/t-5f872409fb2c3900a82e109e-intrinsic-and-molecular-subtypes-of-breast-cancer">Intrinsic and Molecular Subtypes of Breast Cancer</a>, by BioRender, October 2020, Copyright 2021 by BioRender.
</p>
</div>
<p>This dataset was selected because it has characteristics of a common use cancer in cancer genomics:</p>
<ul>
<li>There are pre-defined subtypes discovered using class discovery approaches (unsupervised methods such as clustering) OR defined by the clinical problem</li>
<li>Subtypes are correlated with differences in clinical outcome and choice of treatment</li>
<li>For the purposes of this lab, it is instructive to use subtypes with known cellular signaling pathways</li>
</ul>
</div>
<div id="labs" class="section level2">
<h2>Labs</h2>
<div id="lab-1-a-gentle-introduction" class="section level3">
<h3>LAB 1: A gentle introduction</h3>
<p>We will classify breast tumours as being one of three subtypes: Luminal A, Luminal B, and Basal-like. For this we will build a <strong>three-way classifier</strong> by integrating four types of -omic data: gene expression, miRNA, DNA methylation and proteomics. The design will be simple, using a single patient similarity network (feature) created from each layer. Similarity will be defined as pairwise Pearson correlation of profiles. Each network will use all measures from a given -omic data type (no filtering or grouping).</p>
<p><img src="images/Lab1_design.jpg" width="874" /></p>
</div>
<div id="lab-2-pathway-features-custom-similarity-metrics" class="section level3">
<h3>LAB 2: Pathway features &amp; custom similarity metrics</h3>
<p>We will build a <strong>binary classifier</strong> of breast tumour subtype using gene expression and clinical data, but will use a slightly more complex design. We will define features at the <em>pathway level</em> for gene expression, so we can score pathways by their ability to predict outcome. With clinical data, we will define features at the level of <em>individual clinical variables</em> (e.g. age). We will also introduce the idea of using custom similarity metrics.</p>
<p><img src="images/Lab2_design.jpg" width="874" /></p>
</div>
</div>
<div id="instructions" class="section level2">
<h2>Instructions</h2>
<ul>
<li>Labs will use <strong>Rstudio</strong>.</li>
<li>Each lab will be provided as blocks of R code you will copy-paste into Rstudio to follow the building of the predictor. Code blocks to be pasted will be in yellow like so.</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode r codeblock"><code class="sourceCode r"><span id="cb1-1"><a href="introduction.html#cb1-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">5</span></span></code></pre></div>
<p>If you hover your mouse over the block, you should see a little clipboard icon. Clicking on that icon will allow you to copy the block in one shot. I would use that.</p>
<p>In each case, the result of executing a code block is shown directly after the code, in grey blocks like the one below. These are not intended to be pasted in.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="introduction.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span></span></code></pre></div>
<p>Sometimes we show “pseudocode” or conceptual code blocks; these will also be in grey. None of these are intended to be pasted in.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="introduction.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># do not paste me in!</span></span>
<span id="cb3-2"><a href="introduction.html#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">loop_over</span> (all my patients){</span>
<span id="cb3-3"><a href="introduction.html#cb3-3" aria-hidden="true" tabindex="-1"></a>    do something useful;</span>
<span id="cb3-4"><a href="introduction.html#cb3-4" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>That’s about it!</p>
<p>Let’s go!</p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body" line-spacing="2">
<div id="ref-Pai2019-iu" class="csl-entry">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">Pai, S. <em>et al.</em> netDx: Interpretable patient classification using integrated patient similarity networks. <em>Mol. Syst. Biol.</em> <strong>15,</strong> e8497 (2019).</div>
</div>
<div id="ref-Pai2021-um" class="csl-entry">
<div class="csl-left-margin">2. </div><div class="csl-right-inline">Pai, S. <em>et al.</em> netDx: Software for building interpretable patient classifiers by multi-’omic data integration using patient similarity networks. <em>F1000Research</em> <strong>9,</strong> 1239 (2021).</div>
</div>
<div id="ref-noauthor_2012-nc" class="csl-entry">
<div class="csl-left-margin">3. </div><div class="csl-right-inline">Comprehensive molecular portraits of human breast tumours. <em>Nature</em> <strong>490,</strong> 61–70 (2012).</div>
</div>
<div id="ref-Yersal2014-km" class="csl-entry">
<div class="csl-left-margin">4. </div><div class="csl-right-inline">Yersal, O. &amp; Barutca, S. Biological subtypes of breast cancer: Prognostic and therapeutic implications. <em>World J. Clin. Oncol.</em> <strong>5,</strong> 412–424 (2014).</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-11-lab-1-three-way-classifier-from-four-genomic-layers.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": false
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/01-Introduction.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
