<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Lab 2: Pathway-level features | Multi-modal Data Integration</title>
  <meta name="description" content="Covers patient classification using multi-modal data integration" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Lab 2: Pathway-level features | Multi-modal Data Integration" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Covers patient classification using multi-modal data integration" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Lab 2: Pathway-level features | Multi-modal Data Integration" />
  
  <meta name="twitter:description" content="Covers patient classification using multi-modal data integration" />
  

<meta name="author" content="Shraddha Pai" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
<link rel="prev" href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"/>

<script src="libs/header-attrs-2.8/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">CBW CAN Multi-modal Data Integration</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Multi-modal Data Integration ( June 2021 )</a></li>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html#class-materials"><i class="fa fa-check"></i>Class materials</a></li>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html#setup-instructions"><i class="fa fa-check"></i>Setup instructions</a></li>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html#basic-programs"><i class="fa fa-check"></i>Basic programs</a>
<ul>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html#cytoscape-installation"><i class="fa fa-check"></i>Cytoscape Installation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><i class="fa fa-check"></i>Lab 1: Build 3-way patient classifier from 4 data layers</a>
<ul>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#intro-1"><i class="fa fa-check"></i>Intro</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#prepare-data"><i class="fa fa-check"></i>Prepare data</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#make-rules-to-convert-data-into-features"><i class="fa fa-check"></i>Make rules to convert data into features</a>
<ul>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#datalist"><i class="fa fa-check"></i>dataList</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#grouplist"><i class="fa fa-check"></i>groupList</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#define-patient-similarity-for-each-network"><i class="fa fa-check"></i>Define patient similarity for each network</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#build-predictor"><i class="fa fa-check"></i>Build predictor</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#examine-output"><i class="fa fa-check"></i>Examine output</a>
<ul>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#get-model-results"><i class="fa fa-check"></i>Get model results</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#conclusion"><i class="fa fa-check"></i>Conclusion</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#sessioninfo"><i class="fa fa-check"></i>sessionInfo</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html"><i class="fa fa-check"></i>Lab 2: Pathway-level features</a>
<ul>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#setup"><i class="fa fa-check"></i>Setup</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#rules-to-create-features-patient-similarity-networks"><i class="fa fa-check"></i>Rules to create features (patient similarity networks)</a>
<ul>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#creating-pathway-level-features-from-rna"><i class="fa fa-check"></i>Creating pathway-level features from RNA</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#individual-clinical-variables-as-features"><i class="fa fa-check"></i>Individual clinical variables as features</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#function-to-create-psn"><i class="fa fa-check"></i>Function to create PSN</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#build-predictor-1"><i class="fa fa-check"></i>Build predictor</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#examine-output-1"><i class="fa fa-check"></i>Examine output</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#view-top-scoring-pathways-with-enrichmentmap"><i class="fa fa-check"></i>View top-scoring pathways with EnrichmentMap</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#sessioninfo-1"><i class="fa fa-check"></i>sessionInfo</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Multi-modal Data Integration</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="lab-2-pathway-level-features" class="section level1">
<h1>Lab 2: Pathway-level features</h1>
<p>In this example, we will use …</p>
<p>ADD FIGURE showing design here.</p>
<p>We also use several strategies and definitions of similarity to create features:</p>
<ul>
<li>Clinical variables: Each <em>variable</em> is its own feature (e.g. age); similarity is defined as <em>normalized difference</em>.</li>
<li>Gene expression: Features are defined at the level of <strong><em>pathways</em></strong>; i.e. a feature groups genes corresponding to the pathway. Similarity is defined as pairwise <em>Pearson correlation</em></li>
</ul>
<div id="setup" class="section level2">
<h2>Setup</h2>
<p>For this example we pull data from the The Cancer Genome Atlas through the BioConductor <code>curatedTCGAData</code> package. The fetch command automatically brings in a <code>MultiAssayExperiment</code> object.</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="lab-2-pathway-level-features.html#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(curatedTCGAData))</span></code></pre></div>
<p>In this call we fetch only the gene expression, proteomic and methylation data; setting <code>dry.run=FALSE</code> initiates the fetching of the data.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="lab-2-pathway-level-features.html#cb70-1" aria-hidden="true" tabindex="-1"></a>brca <span class="ot">&lt;-</span> <span class="fu">suppressMessages</span>(</span>
<span id="cb70-2"><a href="lab-2-pathway-level-features.html#cb70-2" aria-hidden="true" tabindex="-1"></a>   <span class="fu">curatedTCGAData</span>(</span>
<span id="cb70-3"><a href="lab-2-pathway-level-features.html#cb70-3" aria-hidden="true" tabindex="-1"></a>       <span class="st">&quot;BRCA&quot;</span>,<span class="fu">c</span>(<span class="st">&quot;mRNAArray&quot;</span>),</span>
<span id="cb70-4"><a href="lab-2-pathway-level-features.html#cb70-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">dry.run=</span><span class="cn">FALSE</span>)</span>
<span id="cb70-5"><a href="lab-2-pathway-level-features.html#cb70-5" aria-hidden="true" tabindex="-1"></a>    )</span></code></pre></div>
<p>Let’s prepare the data:</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="lab-2-pathway-level-features.html#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;prepare_data.R&quot;</span>)</span>
<span id="cb71-2"><a href="lab-2-pathway-level-features.html#cb71-2" aria-hidden="true" tabindex="-1"></a>brca <span class="ot">&lt;-</span> <span class="fu">prepareDataForCBW</span>(brca)</span></code></pre></div>
<pre><code>## harmonizing input:
##   removing 59 sampleMap rows with &#39;colname&#39; not in colnames of experiments</code></pre>
</div>
<div id="rules-to-create-features-patient-similarity-networks" class="section level2">
<h2>Rules to create features (patient similarity networks)</h2>
<p>Load the <code>netDx</code> package.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="lab-2-pathway-level-features.html#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressWarnings</span>(<span class="fu">suppressMessages</span>(<span class="fu">require</span>(netDx)))</span></code></pre></div>
<p>Our plan is to group gene expression data by pathways and clinical data by single variables.</p>
<p>In the code below, we fetch pathway definitions for March 2021 from (<a href="http://download.baderlab.org/EM_Genesets" class="uri">http://download.baderlab.org/EM_Genesets</a>) and group gene expression data by pathways.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="lab-2-pathway-level-features.html#cb74-1" aria-hidden="true" tabindex="-1"></a>groupList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span></code></pre></div>
<div id="creating-pathway-level-features-from-rna" class="section level3">
<h3>Creating pathway-level features from RNA</h3>
<p>Let’s group genes into <em>pathway-level features</em>, i.e. instead of one PSN for transcriptomic data, we create one PSN for each <em>pathway</em>, using transcriptomic data. So if you had a pathway set with 2,000 pathways, this would generate 2,000 input PSN.</p>
<p>There are two ways of providing pathway data to netDx:
1. You can <strong>download</strong> a compilation of pathways from curated databases using the <code>fetchPathwayDefinitions()</code> function in netDx, like so:</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="lab-2-pathway-level-features.html#cb75-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">fetchPathwayDefinitions</span>(<span class="st">&quot;March&quot;</span>,<span class="dv">2021</span>)</span></code></pre></div>
<pre><code>## Fetching http://download.baderlab.org/EM_Genesets/March_01_2021/Human/symbol/Human_AllPathways_March_01_2021_symbol.gmt</code></pre>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="lab-2-pathway-level-features.html#cb77-1" aria-hidden="true" tabindex="-1"></a>x</span></code></pre></div>
<pre><code>##                                                                     BFC2 
## &quot;~/.cache/netDx/20e22fee64ce_Human_AllPathways_March_01_2021_symbol.gmt&quot;</code></pre>
<p>The pathway file was downloaded from <a href="http://download.baderlab.org/EM_Genesets">download.baderlab.org/EM_Genesets</a>, which contains pathways compiled from curated pathway databases such as Reactome, Panther, NCI, MSigDB, etc., [REF], which is updated monthly. Data is contained in GMT format, a common format to represent gene-sets, such as pathways.</p>
<p>We can look at the first 20 lines of the file to see it:</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="lab-2-pathway-level-features.html#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co">#system2(sprintf(&quot;head -n 20 %s&quot;, x))</span></span></code></pre></div>
<p>We then use <code>readPathways()</code> to read the pathways into a list format to provide the predictor with:</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="lab-2-pathway-level-features.html#cb80-1" aria-hidden="true" tabindex="-1"></a>pathList <span class="ot">&lt;-</span> <span class="fu">readPathways</span>(<span class="fu">fetchPathwayDefinitions</span>(<span class="st">&quot;March&quot;</span>,<span class="dv">2021</span>))</span>
<span id="cb80-2"><a href="lab-2-pathway-level-features.html#cb80-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pathList)</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Alternatively, you can also provide a custom pathway set to netDx by reading in a GMT file, using the <code>readPathways()</code> function. In this example, I’ve downloaded a geneset of <a href="http://www.gsea-msigdb.org/gsea/msigdb/collections.jsp#C7">pathways often dysregulated in cancer</a>, set C6 from MSigDB (REF):</li>
</ol>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="lab-2-pathway-level-features.html#cb81-1" aria-hidden="true" tabindex="-1"></a>myGMTfile <span class="ot">&lt;-</span> <span class="st">&quot;supporting_files/c6.all.c7.4.symbols.gmt&quot;</span></span>
<span id="cb81-2"><a href="lab-2-pathway-level-features.html#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="co">#x &lt;- readPathways(myGMTfile)</span></span>
<span id="cb81-3"><a href="lab-2-pathway-level-features.html#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="co">#head(x)</span></span></code></pre></div>
<p>For this tutorial we will limit ourselves to the first set of pathways and use <code>groupList</code> to tell netDx to group transcriptomic data using pathways.</p>
<p><strong>NOTE: The pathway definition file should use the same identifier type as your data. For instance, if the genes in your transcriptomic data are represented using HGNC symbols, then your pathway definition file should also use HGNC symbols (e.g. <em>ID2S</em>) and not a different type of identifier, such as Ensembl IDs (which look like this: <em>ENSG00000010404</em>).</strong></p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="lab-2-pathway-level-features.html#cb82-1" aria-hidden="true" tabindex="-1"></a>groupList[[<span class="st">&quot;BRCA_mRNAArray-20160128&quot;</span>]] <span class="ot">&lt;-</span> pathList[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code></pre></div>
</div>
</div>
<div id="individual-clinical-variables-as-features" class="section level2">
<h2>Individual clinical variables as features</h2>
<p>Models often include clinical variables such as demographic or disease-related features such as age, sex, or treatment regimen. In this example, we take two variables from the sample metadata, and include create one PSN <em>for each variable</em>.</p>
<p>So here, we map the following:</p>
<ul>
<li><code>patient.age_at_initial_pathologic_diagnosis</code> to <code>age</code> (just shorter)</li>
<li><code>stage</code> to <code>STAGE</code></li>
</ul>
<p>These variables must be present in the <code>colData()</code> slot:</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="lab-2-pathway-level-features.html#cb83-1" aria-hidden="true" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> <span class="fu">colData</span>(brca)</span>
<span id="cb83-2"><a href="lab-2-pathway-level-features.html#cb83-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pheno[,<span class="fu">c</span>(<span class="st">&quot;patient.age_at_initial_pathologic_diagnosis&quot;</span>,<span class="st">&quot;stage&quot;</span>)])</span></code></pre></div>
<p>We now add the entry into <code>groupList</code>, which - recall - is where you tell netDx how to group variables. The entry for <code>clinical</code> is special because netDx will look for corresponding variables in the sample metadata table, <code>colData()</code>, rather than in a separate layer in <code>assays()</code>.</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="lab-2-pathway-level-features.html#cb84-1" aria-hidden="true" tabindex="-1"></a>groupList[[<span class="st">&quot;clinical&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb84-2"><a href="lab-2-pathway-level-features.html#cb84-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">age=</span><span class="st">&quot;patient.age_at_initial_pathologic_diagnosis&quot;</span>,</span>
<span id="cb84-3"><a href="lab-2-pathway-level-features.html#cb84-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">stage=</span><span class="st">&quot;STAGE&quot;</span></span>
<span id="cb84-4"><a href="lab-2-pathway-level-features.html#cb84-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="function-to-create-psn" class="section level2">
<h2>Function to create PSN</h2>
<p>In this example, we will introduce the idea of different similarity metrics for gene expression vs clinical data, and show you how to provide a custom similarity function to <code>netDx</code>.</p>
<p>This is <code>makeNets()</code> just as we had seen in the previous exercise, which provides <code>netDx</code> with a custom function to generate similarity networks (i.e. features). We previously used the following code to create PSN based on Pearson correlation:</p>
<p><em>(Do not paste this in)</em></p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="lab-2-pathway-level-features.html#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">makePSN_NamedMatrix</span>(..., <span class="at">writeProfiles=</span><span class="cn">TRUE</span>,...)<span class="st">`</span></span></code></pre></div>
<p>We will now make a different call to <code>makePSN_NamedMatrix()</code> but this time, requesting the use of the normalized difference similarity metric.</p>
<p>This is achieved by calling the following code <em>(Do not paste this in, either)</em>:</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="lab-2-pathway-level-features.html#cb86-1" aria-hidden="true" tabindex="-1"></a>   <span class="fu">makePSN_NamedMatrix</span>(,..., </span>
<span id="cb86-2"><a href="lab-2-pathway-level-features.html#cb86-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">simMetric=</span><span class="st">&quot;custom&quot;</span>, <span class="at">customFunc=</span>normDiff,</span>
<span id="cb86-3"><a href="lab-2-pathway-level-features.html#cb86-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">writeProfiles=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p><code>normDiff</code> is a function provided in the <code>netDx</code> package, but the user may define custom similarity functions in this block of code and pass those to <code>makePSN_NamedMatrix()</code>, using the <code>customFunc</code> parameters; additionally set <code>simMetric</code> to <code>custom</code>.</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="lab-2-pathway-level-features.html#cb87-1" aria-hidden="true" tabindex="-1"></a>makeNets <span class="ot">&lt;-</span> <span class="cf">function</span>(dataList, groupList, netDir,...) {</span>
<span id="cb87-2"><a href="lab-2-pathway-level-features.html#cb87-2" aria-hidden="true" tabindex="-1"></a>    netList <span class="ot">&lt;-</span> <span class="fu">c</span>() </span>
<span id="cb87-3"><a href="lab-2-pathway-level-features.html#cb87-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (nm <span class="cf">in</span> <span class="fu">setdiff</span>(<span class="fu">names</span>(groupList),<span class="st">&quot;clinical&quot;</span>)) {</span>
<span id="cb87-4"><a href="lab-2-pathway-level-features.html#cb87-4" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(groupList[[nm]])) { <span class="do">## REMEMBER TO CHECK FOR NULL</span></span>
<span id="cb87-5"><a href="lab-2-pathway-level-features.html#cb87-5" aria-hidden="true" tabindex="-1"></a>        netList <span class="ot">&lt;-</span> <span class="fu">makePSN_NamedMatrix</span>(</span>
<span id="cb87-6"><a href="lab-2-pathway-level-features.html#cb87-6" aria-hidden="true" tabindex="-1"></a>            dataList[[nm]],</span>
<span id="cb87-7"><a href="lab-2-pathway-level-features.html#cb87-7" aria-hidden="true" tabindex="-1"></a>            <span class="fu">rownames</span>(dataList[[nm]]),</span>
<span id="cb87-8"><a href="lab-2-pathway-level-features.html#cb87-8" aria-hidden="true" tabindex="-1"></a>            groupList[[nm]],</span>
<span id="cb87-9"><a href="lab-2-pathway-level-features.html#cb87-9" aria-hidden="true" tabindex="-1"></a>            netDir,</span>
<span id="cb87-10"><a href="lab-2-pathway-level-features.html#cb87-10" aria-hidden="true" tabindex="-1"></a>            <span class="at">verbose=</span><span class="cn">FALSE</span>,</span>
<span id="cb87-11"><a href="lab-2-pathway-level-features.html#cb87-11" aria-hidden="true" tabindex="-1"></a>            <span class="at">writeProfiles=</span><span class="cn">TRUE</span>,</span>
<span id="cb87-12"><a href="lab-2-pathway-level-features.html#cb87-12" aria-hidden="true" tabindex="-1"></a>            ...) </span>
<span id="cb87-13"><a href="lab-2-pathway-level-features.html#cb87-13" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb87-14"><a href="lab-2-pathway-level-features.html#cb87-14" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb87-15"><a href="lab-2-pathway-level-features.html#cb87-15" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb87-16"><a href="lab-2-pathway-level-features.html#cb87-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># make clinical nets (normalized difference)</span></span>
<span id="cb87-17"><a href="lab-2-pathway-level-features.html#cb87-17" aria-hidden="true" tabindex="-1"></a>    netList2 <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb87-18"><a href="lab-2-pathway-level-features.html#cb87-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(groupList[[<span class="st">&quot;clinical&quot;</span>]])) {</span>
<span id="cb87-19"><a href="lab-2-pathway-level-features.html#cb87-19" aria-hidden="true" tabindex="-1"></a>    netList2 <span class="ot">&lt;-</span> <span class="fu">makePSN_NamedMatrix</span>(</span>
<span id="cb87-20"><a href="lab-2-pathway-level-features.html#cb87-20" aria-hidden="true" tabindex="-1"></a>        dataList<span class="sc">$</span>clinical, </span>
<span id="cb87-21"><a href="lab-2-pathway-level-features.html#cb87-21" aria-hidden="true" tabindex="-1"></a>        <span class="fu">rownames</span>(dataList<span class="sc">$</span>clinical),</span>
<span id="cb87-22"><a href="lab-2-pathway-level-features.html#cb87-22" aria-hidden="true" tabindex="-1"></a>        groupList[[<span class="st">&quot;clinical&quot;</span>]],netDir,</span>
<span id="cb87-23"><a href="lab-2-pathway-level-features.html#cb87-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">simMetric=</span><span class="st">&quot;custom&quot;</span>,<span class="at">customFunc=</span>normDiff, <span class="do">### Notice simMetric &amp; customFunc</span></span>
<span id="cb87-24"><a href="lab-2-pathway-level-features.html#cb87-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">writeProfiles=</span><span class="cn">FALSE</span>,</span>
<span id="cb87-25"><a href="lab-2-pathway-level-features.html#cb87-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">sparsify=</span><span class="cn">TRUE</span>,<span class="at">verbose=</span><span class="cn">TRUE</span>,...)</span>
<span id="cb87-26"><a href="lab-2-pathway-level-features.html#cb87-26" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb87-27"><a href="lab-2-pathway-level-features.html#cb87-27" aria-hidden="true" tabindex="-1"></a>    netList <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">unlist</span>(netList),<span class="fu">unlist</span>(netList2))</span>
<span id="cb87-28"><a href="lab-2-pathway-level-features.html#cb87-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(netList)</span>
<span id="cb87-29"><a href="lab-2-pathway-level-features.html#cb87-29" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="build-predictor-1" class="section level2">
<h2>Build predictor</h2>
<p>Finally!
We have:</p>
<ul>
<li>prepared our data,</li>
<li>grouped RNA by pathways (<code>fetchPathwayDefinitions()</code>, <code>readPathways()</code>),</li>
<li>created two PSN using clinical variables (<code>groupList$clincial</code>), and</li>
<li>defined our similarity function (<code>makeNets()</code>).</li>
</ul>
<p>Now we build our predictor.s</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="lab-2-pathway-level-features.html#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>) <span class="co"># make results reproducible</span></span>
<span id="cb88-2"><a href="lab-2-pathway-level-features.html#cb88-2" aria-hidden="true" tabindex="-1"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="fu">tempdir</span>(),<span class="st">&quot;pred_output&quot;</span>,<span class="at">sep=</span><span class="fu">getFileSep</span>()) <span class="co"># use absolute path</span></span>
<span id="cb88-3"><a href="lab-2-pathway-level-features.html#cb88-3" aria-hidden="true" tabindex="-1"></a>numSplits <span class="ot">&lt;-</span> 2L</span>
<span id="cb88-4"><a href="lab-2-pathway-level-features.html#cb88-4" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">suppressMessages</span>(</span>
<span id="cb88-5"><a href="lab-2-pathway-level-features.html#cb88-5" aria-hidden="true" tabindex="-1"></a>   <span class="fu">buildPredictor</span>(</span>
<span id="cb88-6"><a href="lab-2-pathway-level-features.html#cb88-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">dataList=</span>brca,</span>
<span id="cb88-7"><a href="lab-2-pathway-level-features.html#cb88-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">groupList=</span>groupList,</span>
<span id="cb88-8"><a href="lab-2-pathway-level-features.html#cb88-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">makeNetFunc=</span>makeNets,</span>
<span id="cb88-9"><a href="lab-2-pathway-level-features.html#cb88-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">outDir=</span>outDir, </span>
<span id="cb88-10"><a href="lab-2-pathway-level-features.html#cb88-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">numSplits=</span>numSplits, </span>
<span id="cb88-11"><a href="lab-2-pathway-level-features.html#cb88-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">featScoreMax=</span>2L, </span>
<span id="cb88-12"><a href="lab-2-pathway-level-features.html#cb88-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">featSelCutoff=</span>1L,</span>
<span id="cb88-13"><a href="lab-2-pathway-level-features.html#cb88-13" aria-hidden="true" tabindex="-1"></a>       <span class="at">numCores=</span>4L</span>
<span id="cb88-14"><a href="lab-2-pathway-level-features.html#cb88-14" aria-hidden="true" tabindex="-1"></a>       )</span>
<span id="cb88-15"><a href="lab-2-pathway-level-features.html#cb88-15" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="examine-output-1" class="section level2">
<h2>Examine output</h2>
<p>As before, we get model results, using <code>getResults()</code> from our helper script, <code>helper.R</code>:</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="lab-2-pathway-level-features.html#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;helper.R&quot;</span>)</span>
<span id="cb89-2"><a href="lab-2-pathway-level-features.html#cb89-2" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">getResults</span>(brca,model,<span class="at">featureSelCutoff=</span>2L,</span>
<span id="cb89-3"><a href="lab-2-pathway-level-features.html#cb89-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">featureSelPct=</span><span class="fl">0.5</span>)</span></code></pre></div>
<p>Let’s examine our confusion matrix:</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="lab-2-pathway-level-features.html#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">&quot;get confusion matrix here&quot;</span>)</span></code></pre></div>
</div>
<div id="view-top-scoring-pathways-with-enrichmentmap" class="section level2">
<h2>View top-scoring pathways with EnrichmentMap</h2>
<p>We will now visualize top-scoring pathways using EnrichmentMap.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="lab-2-pathway-level-features.html#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="do">###plotEmap(gmtFiles[[1]],nodeAttrFiles[[1]],</span></span>
<span id="cb91-2"><a href="lab-2-pathway-level-features.html#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="do">###         groupClusters=TRUE, hideNodeLabels=TRUE)</span></span></code></pre></div>
</div>
<div id="sessioninfo-1" class="section level2">
<h2>sessionInfo</h2>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="lab-2-pathway-level-features.html#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": false
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/03-PathwayFeatures.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
