<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Lab 1: Build 3-way patient classifier from 4 data layers | Multi-modal Data Integration</title>
  <meta name="description" content="Covers patient classification using multi-modal data integration" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Lab 1: Build 3-way patient classifier from 4 data layers | Multi-modal Data Integration" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Covers patient classification using multi-modal data integration" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Lab 1: Build 3-way patient classifier from 4 data layers | Multi-modal Data Integration" />
  
  <meta name="twitter:description" content="Covers patient classification using multi-modal data integration" />
  

<meta name="author" content="Shraddha Pai" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
<link rel="prev" href="intro.html"/>
<link rel="next" href="lab-2-pathway-level-features.html"/>
<script src="libs/header-attrs-2.8/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">CBW CAN Multi-modal Data Integration</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Multi-modal Data Integration ( June 2021 )</a></li>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html#class-materials"><i class="fa fa-check"></i>Class materials</a></li>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html#setup-instructions"><i class="fa fa-check"></i>Setup instructions</a></li>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html#basic-programs"><i class="fa fa-check"></i>Basic programs</a>
<ul>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html#cytoscape-installation"><i class="fa fa-check"></i>Cytoscape Installation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><i class="fa fa-check"></i>Lab 1: Build 3-way patient classifier from 4 data layers</a>
<ul>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#intro-1"><i class="fa fa-check"></i>Intro</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#prepare-data"><i class="fa fa-check"></i>Prepare data</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#make-rules-to-convert-data-into-features"><i class="fa fa-check"></i>Make rules to convert data into features</a>
<ul>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#datalist"><i class="fa fa-check"></i>dataList</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#grouplist"><i class="fa fa-check"></i>groupList</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#define-patient-similarity-for-each-network"><i class="fa fa-check"></i>Define patient similarity for each network</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#build-predictor"><i class="fa fa-check"></i>Build predictor</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#examine-output"><i class="fa fa-check"></i>Examine output</a>
<ul>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#get-model-results"><i class="fa fa-check"></i>Get model results</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#conclusion"><i class="fa fa-check"></i>Conclusion</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#sessioninfo"><i class="fa fa-check"></i>sessionInfo</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html"><i class="fa fa-check"></i>Lab 2: Pathway-level features</a>
<ul>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#setup"><i class="fa fa-check"></i>Setup</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#rules-to-create-features-patient-similarity-networks"><i class="fa fa-check"></i>Rules to create features (patient similarity networks)</a>
<ul>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#creating-pathway-level-features-from-rna"><i class="fa fa-check"></i>Creating pathway-level features from RNA</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#individual-clinical-variables-as-features"><i class="fa fa-check"></i>Individual clinical variables as features</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#function-to-create-psn"><i class="fa fa-check"></i>Function to create PSN</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#build-predictor-1"><i class="fa fa-check"></i>Build predictor</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#examine-output-1"><i class="fa fa-check"></i>Examine output</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#view-top-scoring-pathways-with-enrichmentmap"><i class="fa fa-check"></i>View top-scoring pathways with EnrichmentMap</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#sessioninfo-1"><i class="fa fa-check"></i>sessionInfo</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Multi-modal Data Integration</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="lab-1-build-3-way-patient-classifier-from-4-data-layers" class="section level1">
<h1>Lab 1: Build 3-way patient classifier from 4 data layers</h1>
<div id="intro-1" class="section level2">
<h2>Intro</h2>
<p>In this lab we will build a predictor to classify three different types of breast tumours, two of a luminal subtype (Luminal A and Luminal B), and one of a basal subtype.</p>
<p>For this we will use data from the The Cancer Genome Atlas, and will integrate four types of -omic data:</p>
<ul>
<li>gene expression (measured on XXX arrays)</li>
<li>DNA methylation (XXX arrays)</li>
<li>proteomic measures from XXX, and</li>
<li>miRNA sequencing</li>
</ul>
</div>
<div id="prepare-data" class="section level2">
<h2>Prepare data</h2>
<p>In this example, we use curated data from The Cancer Genome Atlas, through the BioConductor <code>curatedTCGAData</code> package. Data for all cancer types profiled in TCGA are available through this package; <a href="https://bioconductor.org/packages/release/data/experiment/vignettes/curatedTCGAData/inst/doc/curatedTCGAData.html">see tutorial for more info</a>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(curatedTCGAData))</span></code></pre></div>
<p>Take a look at the available data without downloading any (set <code>dry.run=TRUE</code>):</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">curatedTCGAData</span>(<span class="at">diseaseCode=</span><span class="st">&quot;BRCA&quot;</span>, <span class="at">assays=</span><span class="st">&quot;*&quot;</span>,<span class="at">dry.run=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## snapshotDate(): 2020-10-27</code></pre>
<pre><code>## See &#39;?curatedTCGAData&#39; for &#39;diseaseCode&#39; and &#39;assays&#39; inputs</code></pre>
<pre><code>##     ah_id                                      title file_size                 rdataclass
## 1   EH584                       BRCA_CNASeq-20160128      0 Mb           RaggedExperiment
## 2   EH585                       BRCA_CNASNP-20160128    9.8 Mb           RaggedExperiment
## 3   EH586                       BRCA_CNVSNP-20160128    2.8 Mb           RaggedExperiment
## 4   EH588             BRCA_GISTIC_AllByGene-20160128    1.3 Mb       SummarizedExperiment
## 5  EH2121                 BRCA_GISTIC_Peaks-20160128      0 Mb RangedSummarizedExperiment
## 6   EH589     BRCA_GISTIC_ThresholdedByGene-20160128    0.4 Mb       SummarizedExperiment
## 7  EH2122  BRCA_Methylation_methyl27-20160128_assays   63.2 Mb       SummarizedExperiment
## 8  EH2123      BRCA_Methylation_methyl27-20160128_se    0.4 Mb       SummarizedExperiment
## 9  EH2124 BRCA_Methylation_methyl450-20160128_assays 2613.2 Mb           RaggedExperiment
## 10 EH2125     BRCA_Methylation_methyl450-20160128_se    6.1 Mb       SummarizedExperiment
## 11  EH593                 BRCA_miRNASeqGene-20160128    0.6 Mb       SummarizedExperiment
## 12  EH594                    BRCA_mRNAArray-20160128   27.3 Mb       SummarizedExperiment
## 13  EH595                     BRCA_Mutation-20160128    4.5 Mb           RaggedExperiment
## 14  EH596              BRCA_RNASeq2GeneNorm-20160128   64.5 Mb       SummarizedExperiment
## 15  EH597                   BRCA_RNASeqGene-20160128     30 Mb       SummarizedExperiment
## 16  EH598                    BRCA_RPPAArray-20160128    1.6 Mb       SummarizedExperiment
##    rdatadateadded rdatadateremoved
## 1      2017-10-10             &lt;NA&gt;
## 2      2017-10-10             &lt;NA&gt;
## 3      2017-10-10             &lt;NA&gt;
## 4      2017-10-10             &lt;NA&gt;
## 5      2019-01-09             &lt;NA&gt;
## 6      2017-10-10             &lt;NA&gt;
## 7      2019-01-09             &lt;NA&gt;
## 8      2019-01-09             &lt;NA&gt;
## 9      2019-01-09             &lt;NA&gt;
## 10     2019-01-09             &lt;NA&gt;
## 11     2017-10-10             &lt;NA&gt;
## 12     2017-10-10             &lt;NA&gt;
## 13     2017-10-10             &lt;NA&gt;
## 14     2017-10-10             &lt;NA&gt;
## 15     2017-10-10             &lt;NA&gt;
## 16     2017-10-10             &lt;NA&gt;</code></pre>
<p>Now let’s actually sdownload the data:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb7-1" aria-hidden="true" tabindex="-1"></a>brca <span class="ot">&lt;-</span> <span class="fu">suppressMessages</span>(<span class="fu">curatedTCGAData</span>(<span class="st">&quot;BRCA&quot;</span>,</span>
<span id="cb7-2"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb7-2" aria-hidden="true" tabindex="-1"></a>                                         <span class="fu">c</span>(<span class="st">&quot;mRNAArray&quot;</span>,<span class="st">&quot;Methylation_methyl27&quot;</span>, </span>
<span id="cb7-3"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb7-3" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">&quot;RPPAArray&quot;</span>,<span class="st">&quot;miRNASeqGene&quot;</span>),</span>
<span id="cb7-4"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb7-4" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">dry.run=</span><span class="cn">FALSE</span>))</span></code></pre></div>
<p>This call returns a <code>MultiAssayExperiment</code> object. Recall that this is a container for storing multiple assays performed on the same set of samples. <a href="https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/QuickStartMultiAssay.html">See this tutorial</a> to learn more.</p>
<p>Let’s briefly explore the <code>brca</code> <code>MultiAssayExperiment</code> object.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb8-1" aria-hidden="true" tabindex="-1"></a>brca</span></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 4 listed
##  experiments with user-defined names and respective classes.
##  Containing an ExperimentList class object of length 4:
##  [1] BRCA_miRNASeqGene-20160128: SummarizedExperiment with 1046 rows and 849 columns
##  [2] BRCA_mRNAArray-20160128: SummarizedExperiment with 17814 rows and 590 columns
##  [3] BRCA_RPPAArray-20160128: SummarizedExperiment with 226 rows and 937 columns
##  [4] BRCA_Methylation_methyl27-20160128: SummarizedExperiment with 27578 rows and 343 columns
## Functionality:
##  experiments() - obtain the ExperimentList instance
##  colData() - the primary/phenotype DataFrame
##  sampleMap() - the sample coordination DataFrame
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
##  *Format() - convert into a long or wide DataFrame
##  assays() - convert ExperimentList to a SimpleList of matrices
##  exportClass() - save all data to files</code></pre>
<p><code>assays()</code> returns a <code>list</code> with all -omic data associated with this object.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">assays</span>(brca))</span></code></pre></div>
<pre><code>## [1] &quot;List object of length 4 with 0 metadata columns&quot;</code></pre>
<p><code>names()</code> shows you what -omic datatypes the layers correspond to:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(<span class="fu">assays</span>(brca))</span></code></pre></div>
<pre><code>## [1] &quot;BRCA_miRNASeqGene-20160128&quot;         &quot;BRCA_mRNAArray-20160128&quot;           
## [3] &quot;BRCA_RPPAArray-20160128&quot;            &quot;BRCA_Methylation_methyl27-20160128&quot;</code></pre>
<p>And here we subset the miRNA data, looking at data for the first five patients</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb14-1" aria-hidden="true" tabindex="-1"></a>mir <span class="ot">&lt;-</span> <span class="fu">assays</span>(brca)[[<span class="st">&quot;BRCA_miRNASeqGene-20160128&quot;</span>]]</span>
<span id="cb14-2"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mir[,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code></pre></div>
<pre><code>##              TCGA-3C-AAAU-01A-11R-A41G-13 TCGA-3C-AALI-01A-11R-A41G-13
## hsa-let-7a-1                        95618                        49201
## hsa-let-7a-2                       189674                        98691
## hsa-let-7a-3                        96815                        49035
## hsa-let-7b                         264034                       148591
## hsa-let-7c                           3641                         5095
## hsa-let-7d                           4333                         3263
##              TCGA-3C-AALJ-01A-31R-A41G-13 TCGA-3C-AALK-01A-11R-A41G-13
## hsa-let-7a-1                        75342                        57278
## hsa-let-7a-2                       150472                       114320
## hsa-let-7a-3                        76206                        57540
## hsa-let-7b                          99938                       164553
## hsa-let-7c                           5799                        18464
## hsa-let-7d                           5658                         2114
##              TCGA-4H-AAAK-01A-12R-A41G-13
## hsa-let-7a-1                        67196
## hsa-let-7a-2                       134563
## hsa-let-7a-3                        67607
## hsa-let-7b                         136918
## hsa-let-7c                          20429
## hsa-let-7d                           2162</code></pre>
<p>Look at sample phenotype data, where rows contain data for each patient and columns have non-omic measures (e.g. clinical data):</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb16-1" aria-hidden="true" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> <span class="fu">colData</span>(brca)</span>
<span id="cb16-2"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pheno)[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]</span></code></pre></div>
<pre><code>##  [1] &quot;patientID&quot;                            &quot;years_to_birth&quot;                      
##  [3] &quot;vital_status&quot;                         &quot;days_to_death&quot;                       
##  [5] &quot;days_to_last_followup&quot;                &quot;tumor_tissue_site&quot;                   
##  [7] &quot;pathologic_stage&quot;                     &quot;pathology_T_stage&quot;                   
##  [9] &quot;pathology_N_stage&quot;                    &quot;pathology_M_stage&quot;                   
## [11] &quot;gender&quot;                               &quot;date_of_initial_pathologic_diagnosis&quot;
## [13] &quot;days_to_last_known_alive&quot;             &quot;radiation_therapy&quot;                   
## [15] &quot;histological_type&quot;                    &quot;number_of_lymph_nodes&quot;               
## [17] &quot;race&quot;                                 &quot;ethnicity&quot;                           
## [19] &quot;admin.bcr&quot;                            &quot;admin.day_of_dcc_upload&quot;</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pheno[,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 5 columns
##                 patientID years_to_birth vital_status days_to_death days_to_last_followup
##               &lt;character&gt;      &lt;integer&gt;    &lt;integer&gt;     &lt;integer&gt;             &lt;integer&gt;
## TCGA-A1-A0SB TCGA-A1-A0SB             70            0            NA                   259
## TCGA-A1-A0SD TCGA-A1-A0SD             59            0            NA                   437
## TCGA-A1-A0SE TCGA-A1-A0SE             56            0            NA                  1321
## TCGA-A1-A0SF TCGA-A1-A0SF             54            0            NA                  1463
## TCGA-A1-A0SG TCGA-A1-A0SG             61            0            NA                   434
## TCGA-A1-A0SH TCGA-A1-A0SH             39            0            NA                  1437</code></pre>
<p>This next code block prepares the TCGA data. This includes:</p>
<ul>
<li>removing duplicate samples</li>
<li>reformatting patient IDs (e.g. removing spaces and hyphens)</li>
<li>creating an <code>ID</code> column in <code>colData(brca)</code>, which contains unique patient IDs</li>
<li>creating a <code>STATUS</code> column in <code>colData(brca)</code> which contains the patient labels (i.e what we want netDx to classify).</li>
</ul>
<p>In practice you would do this once and save the data before running netDx. This code has been moved into a supporting file, <code>prepare_data.R</code>. You can explore it after the lab to see how some things are achieved (e.g. removing duplicate samples).</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;prepare_data.R&quot;</span>)</span>
<span id="cb20-2"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb20-2" aria-hidden="true" tabindex="-1"></a>brca <span class="ot">&lt;-</span> <span class="fu">prepareDataForCBW</span>(brca)</span></code></pre></div>
<pre><code>## harmonizing input:
##   removing 42 sampleMap rows with &#39;colname&#39; not in colnames of experiments</code></pre>
<pre><code>## harmonizing input:
##   removing 59 sampleMap rows with &#39;colname&#39; not in colnames of experiments</code></pre>
<pre><code>## harmonizing input:
##   removing 19 sampleMap rows with &#39;colname&#39; not in colnames of experiments</code></pre>
<pre><code>## harmonizing input:
##   removing 26 sampleMap rows with &#39;colname&#39; not in colnames of experiments</code></pre>
<p>Notice that we now have <code>ID</code> and <code>STATUS</code> columns in the sample metadata table.
<strong>Note: These columns are required by netDx. </strong></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb25-1" aria-hidden="true" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> <span class="fu">colData</span>(brca)</span>
<span id="cb25-2"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pheno[,<span class="fu">c</span>(<span class="st">&quot;ID&quot;</span>,<span class="st">&quot;STATUS&quot;</span>)])</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 2 columns
##                        ID      STATUS
##               &lt;character&gt; &lt;character&gt;
## TCGA-A1-A0SD TCGA-A1-A0SD   Luminal.A
## TCGA-A1-A0SE TCGA-A1-A0SE   Luminal.A
## TCGA-A1-A0SH TCGA-A1-A0SH   Luminal.A
## TCGA-A1-A0SJ TCGA-A1-A0SJ   Luminal.A
## TCGA-A1-A0SK TCGA-A1-A0SK  Basal.like
## TCGA-A1-A0SM TCGA-A1-A0SM   Luminal.B</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(pheno<span class="sc">$</span>STATUS,<span class="at">useNA=</span><span class="st">&quot;always&quot;</span>)  <span class="co"># useNA lets you see </span></span></code></pre></div>
<pre><code>## 
## Basal.like  Luminal.A  Luminal.B       &lt;NA&gt; 
##         97        230        127          0</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb29-1" aria-hidden="true" tabindex="-1"></a>                                    <span class="co"># unintentional missing values</span></span></code></pre></div>
</div>
<div id="make-rules-to-convert-data-into-features" class="section level2">
<h2>Make rules to convert data into features</h2>
<p>Now let’s set up the data for input to netDx.
netDx allows the user to define how data is converted into patient similarity networks (or PSNs), which are the features that go into the model. This is done specifically by telling the model how to group different types of data and how to define similarity for each of these (e.g. Pearson correlation, normalized difference, etc.,).</p>
<p>The relevant input parameters are:</p>
<ul>
<li><code>dataList</code>: the patient data, provided as a <code>MultiAssayExperiment</code> object. Refer to the <a href="https://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html">tutorials for MultiAssayExperiment</a> to see how to construct those objects from data.</li>
<li><code>groupList</code>: sets of input data that would correspond to individual networks (e.g. genes grouped into pathways)</li>
</ul>
<p>Let’s start by loading the <code>netDx</code> package.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressWarnings</span>(<span class="fu">suppressMessages</span>(<span class="fu">require</span>(netDx)))</span></code></pre></div>
<p>Let’s set up each of the input arguments one by one.</p>
<div id="datalist" class="section level3">
<h3>dataList</h3>
<p><strong>What is this:</strong> The input data.</p>
<p><strong>Format:</strong> <code>MultiAssayExperiment</code> object.</p>
<p>We’ve already set this up; this is the <code>brca</code> object, so nothing more to do here.</p>
</div>
<div id="grouplist" class="section level3">
<h3>groupList</h3>
<p><strong>What is this:</strong> <code>groupList</code> tells netDx how to group measures within a layer, to generate a PSN. Measures could be individual genes, proteins, CpG bases (in DNA methylation data), clinical variables, etc.,</p>
<p>In this simple example we just create one PSN per datatype:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb31-1" aria-hidden="true" tabindex="-1"></a>expr <span class="ot">&lt;-</span> <span class="fu">assays</span>(brca)</span>
<span id="cb31-2"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb31-2" aria-hidden="true" tabindex="-1"></a>groupList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb31-3"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(expr)) {</span>
<span id="cb31-4"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb31-4" aria-hidden="true" tabindex="-1"></a>    cur <span class="ot">&lt;-</span> expr[[k]]; nm <span class="ot">&lt;-</span> <span class="fu">names</span>(expr)[k]</span>
<span id="cb31-5"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb31-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># all measure names should be in rownames column</span></span>
<span id="cb31-6"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb31-6" aria-hidden="true" tabindex="-1"></a>    groupList[[nm]] <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">nm=</span><span class="fu">rownames</span>(cur)) </span>
<span id="cb31-7"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb31-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(groupList[[nm]])[<span class="dv">1</span>] <span class="ot">&lt;-</span> nm;</span>
<span id="cb31-8"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb31-8" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p><code>groupList</code> is a two tiered list, or list-of-lists. The first tier is for each data layers, with names matching those in <code>assays(brca)</code>. The second tier contains all the PSNs we want to make for that layer.</p>
<p>**SP:</p>
<p>Here is the first layer:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(groupList)</span></code></pre></div>
<pre><code>##                                    Length Class  Mode
## BRCA_miRNASeqGene-20160128         1      -none- list
## BRCA_mRNAArray-20160128            1      -none- list
## BRCA_RPPAArray-20160128            1      -none- list
## BRCA_Methylation_methyl27-20160128 1      -none- list</code></pre>
<p>In this lab exercise we create only one PSN per data layer, using all the measures from an -omic assay. e.g. One PSN based on similarity across entire transcriptome, one for methylome, etc.,). So in this lab exercise, the inner tier simply contains one entry, with all measures for the given layer. This design will get more interested in Lab 2, when we create pathway-level features.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(groupList[[<span class="st">&quot;BRCA_mRNAArray-20160128&quot;</span>]])</span></code></pre></div>
<pre><code>## [1] &quot;BRCA_mRNAArray-20160128&quot;</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(groupList[[<span class="st">&quot;BRCA_mRNAArray-20160128&quot;</span>]][[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>## [1] &quot;ELMO2&quot;    &quot;CREB3L1&quot;  &quot;RPS11&quot;    &quot;PNMA1&quot;    &quot;MMP2&quot;     &quot;C10orf90&quot;</code></pre>
</div>
<div id="define-patient-similarity-for-each-network" class="section level3">
<h3>Define patient similarity for each network</h3>
<p>This function is defined by the user and tells the predictor how to create networks from the provided input data.</p>
<p>This function requires <code>dataList</code>,<code>groupList</code>, and <code>netDir</code> as input variables. The residual <code>...</code> parameter is to pass additional variables to <code>makePSN_NamedMatrix()</code>, notably <code>numCores</code> (number of parallel jobs).</p>
<p>netDx requires that this function have:</p>
<ul>
<li><code>dataList</code>,<code>groupList</code>, and <code>netDir</code> as input variables. The residual <code>...</code> parameter is to pass additional variables to <code>makePSN_NamedMatrix()</code>, notably number of cores for parallel processing (<code>numCores</code>).</li>
<li></li>
</ul>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-1" aria-hidden="true" tabindex="-1"></a>makeNets <span class="ot">&lt;-</span> <span class="cf">function</span>(dataList, groupList, netDir,...) {</span>
<span id="cb38-2"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-2" aria-hidden="true" tabindex="-1"></a>    netList <span class="ot">&lt;-</span> <span class="fu">c</span>() <span class="co"># initialize before is.null() check</span></span>
<span id="cb38-3"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb38-4"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-4" aria-hidden="true" tabindex="-1"></a>    layerNames <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;BRCA_miRNASeqGene-20160128&quot;</span>,</span>
<span id="cb38-5"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">&quot;BRCA_mRNAArray-20160128&quot;</span>,</span>
<span id="cb38-6"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">&quot;BRCA_RPPAArray-20160128&quot;</span>,</span>
<span id="cb38-7"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">&quot;BRCA_Methylation_methyl27-20160128&quot;</span>)</span>
<span id="cb38-8"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb38-9"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (nm <span class="cf">in</span> layerNames){             <span class="do">## for each layer</span></span>
<span id="cb38-10"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-10" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(groupList[[nm]])){ <span class="do">## must check for null for each layer</span></span>
<span id="cb38-11"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-11" aria-hidden="true" tabindex="-1"></a>            netList_cur <span class="ot">&lt;-</span> <span class="fu">makePSN_NamedMatrix</span>(</span>
<span id="cb38-12"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-12" aria-hidden="true" tabindex="-1"></a>                dataList[[nm]],</span>
<span id="cb38-13"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-13" aria-hidden="true" tabindex="-1"></a>                <span class="fu">rownames</span>(dataList[[nm]]),   <span class="do">## names of measures (e.g. genes, CpGs)</span></span>
<span id="cb38-14"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-14" aria-hidden="true" tabindex="-1"></a>                groupList[[nm]],            <span class="do">## how to group measures in that layer</span></span>
<span id="cb38-15"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-15" aria-hidden="true" tabindex="-1"></a>                netDir,                     <span class="do">## leave this as-is, netDx will figure out where this is.</span></span>
<span id="cb38-16"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-16" aria-hidden="true" tabindex="-1"></a>                <span class="at">verbose=</span><span class="cn">TRUE</span>,           </span>
<span id="cb38-17"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-17" aria-hidden="true" tabindex="-1"></a>                <span class="at">writeProfiles=</span><span class="cn">TRUE</span>,         <span class="do">## use Pearson correlation-based similarity</span></span>
<span id="cb38-18"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-18" aria-hidden="true" tabindex="-1"></a>                ...</span>
<span id="cb38-19"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-19" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb38-20"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-21"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-21" aria-hidden="true" tabindex="-1"></a>            netList <span class="ot">&lt;-</span> <span class="fu">c</span>(netList,netList_cur)   <span class="do">## just leave this in</span></span>
<span id="cb38-22"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-22" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb38-23"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-23" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb38-24"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">unlist</span>(netList)) <span class="do">## just leave this in </span></span>
<span id="cb38-25"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb38-25" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p><strong><em>Note:</em></strong> While netDx provides a high degree of flexibility in achieving your design of choice, it is up to the user to ensure that the design, i.e. the similarity metric and variable groupings, is appropriate for your application. Domain knowledge is almost likely required for good design.</p>
</div>
</div>
<div id="build-predictor" class="section level2">
<h2>Build predictor</h2>
<p>Finally we call the function that runs the netDx predictor. We provide:</p>
<ul>
<li>patient data (<code>dataList</code>)</li>
<li>grouping rules (<code>groupList</code>)</li>
<li>function to create PSN from data, includes choice of similarity metric (<code>makeNetFunc</code>)</li>
<li>number of train/test splits over which to collect feature scores and average performance: <code>numSplits</code>,</li>
<li>maximum score for features in one round of feature selection (<code>featScoreMax</code>, set to 10)</li>
<li>threshold to call feature-selected networks for each train/test split (<code>featSelCutoff</code>); only features scoring this value or higher will be used to classify test patients,</li>
<li>number of cores to use for parallel processing (<code>numCores</code>).</li>
</ul>
<p>The call below runs 2 train/test splits, to be able to run fast.
Within each split, it:</p>
<ul>
<li>splits data into train/test using the default split of 80:20</li>
<li>score2 networks between 0 to 2 (i.e. <code>featScoreMax=2</code>)</li>
<li>uses networks that score &gt;=1 out of 2 (<code>featSelCutoff</code>) to classify test samples for that split.</li>
</ul>
<p>These are unrealistically low values set so the example will run fast. In practice a good starting point is <code>featScoreMax=10</code>, <code>featSelCutoff=9</code> and <code>numSplits=100</code>, but these parameters depend on the sample sizes in the dataset and heterogeneity of the samples.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>) <span class="co"># make results reproducible</span></span>
<span id="cb39-2"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-2" aria-hidden="true" tabindex="-1"></a>outDir <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="fu">tempdir</span>(),<span class="fu">randAlphanumString</span>(),</span>
<span id="cb39-3"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;pred_output&quot;</span>,<span class="at">sep=</span><span class="fu">getFileSep</span>())</span>
<span id="cb39-4"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="do">###if (file.exists(outDir)) unlink(outDir,recursive=TRUE)</span></span>
<span id="cb39-5"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="do">###</span></span>
<span id="cb39-6"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-6" aria-hidden="true" tabindex="-1"></a><span class="do">###model &lt;- buildPredictor(</span></span>
<span id="cb39-7"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="do">###      dataList=brca,     ## your data</span></span>
<span id="cb39-8"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="do">###   groupList=groupList,  ## grouping strategy</span></span>
<span id="cb39-9"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-9" aria-hidden="true" tabindex="-1"></a><span class="do">###      makeNetFunc=makeNets,  ## function to build PSNs</span></span>
<span id="cb39-10"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-10" aria-hidden="true" tabindex="-1"></a><span class="do">###    outDir=outDir,       ## output directory</span></span>
<span id="cb39-11"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-11" aria-hidden="true" tabindex="-1"></a><span class="do">###   trainProp=0.8,        ## pct of samples to use to train model in</span></span>
<span id="cb39-12"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-12" aria-hidden="true" tabindex="-1"></a>                            <span class="do">## each split</span></span>
<span id="cb39-13"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-13" aria-hidden="true" tabindex="-1"></a><span class="do">###      numSplits=2L,          ## number of train/test splits</span></span>
<span id="cb39-14"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-14" aria-hidden="true" tabindex="-1"></a><span class="do">###   featSelCutoff=1L,     ## threshold for calling something</span></span>
<span id="cb39-15"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-15" aria-hidden="true" tabindex="-1"></a>                            <span class="do">## feature-selected</span></span>
<span id="cb39-16"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-16" aria-hidden="true" tabindex="-1"></a><span class="do">###   featScoreMax=2L,      ## max score for feature selection</span></span>
<span id="cb39-17"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-17" aria-hidden="true" tabindex="-1"></a><span class="do">###      numCores=4L,           ## set higher for parallelizing</span></span>
<span id="cb39-18"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-18" aria-hidden="true" tabindex="-1"></a><span class="do">###   debugMode=FALSE,</span></span>
<span id="cb39-19"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-19" aria-hidden="true" tabindex="-1"></a><span class="do">###   keepAllData=FALSE,    ## set to TRUE for debugging or low-level files used by the predictor</span></span>
<span id="cb39-20"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-20" aria-hidden="true" tabindex="-1"></a><span class="do">###      logging=&quot;none&quot;</span></span>
<span id="cb39-21"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-21" aria-hidden="true" tabindex="-1"></a><span class="do">###   )</span></span>
<span id="cb39-22"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb39-22" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;/home/spai/data/CBW2021/model_twosplits.rda&quot;</span>)</span></code></pre></div>
</div>
<div id="examine-output" class="section level2">
<h2>Examine output</h2>
<p>The results are stored in the list object returned by the <code>buildPredictor()</code> call.
This list contains:</p>
<ul>
<li><code>inputNets</code>: all input networks that the model started with.</li>
<li><code>Split&lt;i&gt;</code>: a list with results for each train-test split
<ul>
<li><code>predictions</code>: real and predicted labels for test patients</li>
<li><code>accuracy</code>: percent accuracy of predictions</li>
<li><code>featureScores</code>: feature scores for each label (list with <code>g</code> entries, where <code>g</code> is number of patient labels). Each entry contains the feature selection scores for the corresponding label.</li>
<li><code>featureSelected</code>: vector of features that pass feature selection. List of length <code>g</code>, with one entry per label.</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model)</span></code></pre></div>
<pre><code>##           Length Class  Mode     
## inputNets 8      -none- character
## Split1    4      -none- list     
## Split2    4      -none- list</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model<span class="sc">$</span>Split1)</span></code></pre></div>
<pre><code>##                 Length Class      Mode   
## featureScores      3   -none-     list   
## featureSelected    3   -none-     list   
## predictions     2692   data.frame list   
## accuracy           1   -none-     numeric</code></pre>
<div id="get-model-results" class="section level3">
<h3>Get model results</h3>
<p>Now we get model output, including performance for various train/test splits and consistently high-scoring features. <code>helper.R</code> contains convenience functions to do this, which will be integrated into upcoming versions of netDx.</p>
<p>In the function below, we define top-scoring features as those which score at least 2 in at least half of the train/test splits</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;helper.R&quot;</span>)</span>
<span id="cb44-2"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb44-2" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">getResults</span>(brca,model,<span class="at">featureSelCutoff=</span>2L,</span>
<span id="cb44-3"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb44-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">featureSelPct=</span><span class="fl">0.5</span>)</span></code></pre></div>
<pre><code>## Detected 2 splits and 3 classes</code></pre>
<pre><code>## * Compiling feature scores and calling selected features</code></pre>
<p><code>results</code> contains <code>performance</code>, <code>selectedFeatures</code> for each patient label, and the table of feature <code>scores</code>.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(results)</span></code></pre></div>
<pre><code>##                  Length Class  Mode
## selectedFeatures 2      -none- list
## featureScores    3      -none- list
## performance      2      -none- list</code></pre>
<p>Look at the performance:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb49-1" aria-hidden="true" tabindex="-1"></a>results<span class="sc">$</span>performance</span></code></pre></div>
<pre><code>## $meanAccuracy
## [1] 0.7282609
## 
## $splitAccuracy
## [1] 0.7391304 0.7173913</code></pre>
<p>Look at feature scores for all labels, across all train-test splits:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb51-1" aria-hidden="true" tabindex="-1"></a>results<span class="sc">$</span>featureScores</span></code></pre></div>
<pre><code>## $Luminal.A
##                              Feature Split1 Split2
## 1 BRCA_Methylation_methyl27-20160128      1      1
## 2            BRCA_RPPAArray-20160128      2      2
## 3            BRCA_mRNAArray-20160128      2      2
## 4         BRCA_miRNASeqGene-20160128      1      1
## 
## $Basal.like
##                              Feature Split1 Split2
## 1 BRCA_Methylation_methyl27-20160128      2      2
## 2            BRCA_RPPAArray-20160128      2      2
## 3            BRCA_mRNAArray-20160128      2      2
## 4         BRCA_miRNASeqGene-20160128      1      2
## 
## $Luminal.B
##                              Feature Split1 Split2
## 1 BRCA_Methylation_methyl27-20160128      2      2
## 2            BRCA_RPPAArray-20160128      2      2
## 3            BRCA_mRNAArray-20160128      2      2</code></pre>
<p>And here are selected features, which are those scoring 2 out of 2 in at least half of the splits. This threshold is simply for illustration. In practice we would run at least 10 train/test splits (ideally 100+), and look for features that score 7+ out of 10 in &gt;70% splits.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb53-1" aria-hidden="true" tabindex="-1"></a>results<span class="sc">$</span>selectedFeatures</span></code></pre></div>
<pre><code>## $featScores
## $featScores$Luminal.A
##                              Feature Split1 Split2
## 1 BRCA_Methylation_methyl27-20160128      1      1
## 2            BRCA_RPPAArray-20160128      2      2
## 3            BRCA_mRNAArray-20160128      2      2
## 4         BRCA_miRNASeqGene-20160128      1      1
## 
## $featScores$Basal.like
##                              Feature Split1 Split2
## 1 BRCA_Methylation_methyl27-20160128      2      2
## 2            BRCA_RPPAArray-20160128      2      2
## 3            BRCA_mRNAArray-20160128      2      2
## 4         BRCA_miRNASeqGene-20160128      1      2
## 
## $featScores$Luminal.B
##                              Feature Split1 Split2
## 1 BRCA_Methylation_methyl27-20160128      2      2
## 2            BRCA_RPPAArray-20160128      2      2
## 3            BRCA_mRNAArray-20160128      2      2
## 
## 
## $selectedFeatures
## $selectedFeatures$Luminal.A
## [1] &quot;BRCA_Methylation_methyl27-20160128&quot; &quot;BRCA_RPPAArray-20160128&quot;           
## [3] &quot;BRCA_mRNAArray-20160128&quot;            &quot;BRCA_miRNASeqGene-20160128&quot;        
## 
## $selectedFeatures$Basal.like
## [1] &quot;BRCA_Methylation_methyl27-20160128&quot; &quot;BRCA_RPPAArray-20160128&quot;           
## [3] &quot;BRCA_mRNAArray-20160128&quot;            &quot;BRCA_miRNASeqGene-20160128&quot;        
## 
## $selectedFeatures$Luminal.B
## [1] &quot;BRCA_Methylation_methyl27-20160128&quot; &quot;BRCA_RPPAArray-20160128&quot;           
## [3] &quot;BRCA_mRNAArray-20160128&quot;</code></pre>
<p>We finally get the integrated PSN and visualize it using a tSNE plot:</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb55-1" aria-hidden="true" tabindex="-1"></a>psn <span class="ot">&lt;-</span> <span class="fu">getPSN</span>(brca,groupList,makeNets,results<span class="sc">$</span>selectedFeatures)</span></code></pre></div>
<pre><code>## BRCA_miRNASeqGene-20160128: 1 features</code></pre>
<pre><code>## BRCA_mRNAArray-20160128: 1 features</code></pre>
<pre><code>## BRCA_RPPAArray-20160128: 1 features</code></pre>
<pre><code>## BRCA_Methylation_methyl27-20160128: 1 features</code></pre>
<pre><code>## * Making integrated PSN</code></pre>
<pre><code>## Warning in dir.create(paste(netDir, &quot;profiles&quot;, sep = fsep)): &#39;/tmp/RtmpH3a8cI/profiles&#39;
## already exists</code></pre>
<pre><code>## Warning in dir.create(paste(netDir, &quot;INTERACTIONS&quot;, sep = fsep)): &#39;/tmp/RtmpH3a8cI/
## INTERACTIONS&#39; already exists</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb63-1" aria-hidden="true" tabindex="-1"></a>tsne <span class="ot">&lt;-</span> <span class="fu">tSNEPlotter</span>(</span>
<span id="cb63-2"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb63-2" aria-hidden="true" tabindex="-1"></a>    psn<span class="sc">$</span>patientSimNetwork_unpruned, </span>
<span id="cb63-3"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb63-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colData</span>(brca)</span>
<span id="cb63-4"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb63-4" aria-hidden="true" tabindex="-1"></a>    )</span></code></pre></div>
<pre><code>## * Making symmetric matrix</code></pre>
<pre><code>## * Running tSNE</code></pre>
<pre><code>## * Plotting</code></pre>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-25-1.png" width="768" /></p>
</div>
</div>
<div id="conclusion" class="section level2">
<h2>Conclusion</h2>
<p>Congratulations! You have successfully classified three types of tumours by integrating four different -omic layers (transcriptomic, methylomic,proteomic and miRNA).</p>
<p>But these features aren’t really interpretable. Let’s change that with our next tutorial.</p>
<p>Best practices side note: We end this tutorial with a call to <code>sessionInfo()</code> which prints the complete environment information for your R session. This is standard output that should be reported to R package managers when you write in with a question, particularly if reporting an error or bug. Sometimes a particular underlying dependency package may be the cause of an error, or you may need to upgrade to a newer version of the package. By including this info in an email, you will allow the other person to better solve your issue.</p>
</div>
<div id="sessioninfo" class="section level2">
<h2>sessionInfo</h2>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.0.5 (2021-03-31)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=C                 LC_NUMERIC=C               LC_TIME=C                 
##  [4] LC_COLLATE=C               LC_MONETARY=C              LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods  
## [9] base     
## 
## other attached packages:
##  [1] plotrix_3.8-1               ggplot2_3.3.3               Rtsne_0.15                 
##  [4] rhdf5_2.34.0                curatedTCGAData_1.12.1      MultiAssayExperiment_1.16.0
##  [7] SummarizedExperiment_1.20.0 GenomicRanges_1.42.0        GenomeInfoDb_1.26.7        
## [10] IRanges_2.24.1              S4Vectors_0.28.1            MatrixGenerics_1.2.1       
## [13] matrixStats_0.58.0          netDx_1.2.3                 bigmemory_4.5.36           
## [16] Biobase_2.50.0              BiocGenerics_0.36.1        
## 
## loaded via a namespace (and not attached):
##   [1] utf8_1.2.1                    R.utils_2.10.1               
##   [3] tidyselect_1.1.1              RSQLite_2.2.7                
##   [5] AnnotationDbi_1.52.0          grid_4.0.5                   
##   [7] combinat_0.0-8                BiocParallel_1.24.1          
##   [9] RNeXML_2.4.5                  munsell_0.5.0                
##  [11] codetools_0.2-18              withr_2.4.2                  
##  [13] colorspace_2.0-1              highr_0.9                    
##  [15] knitr_1.33                    uuid_0.1-4                   
##  [17] zinbwave_1.12.0               rstudioapi_0.13              
##  [19] SingleCellExperiment_1.12.0   ROCR_1.0-11                  
##  [21] NMF_0.23.0                    labeling_0.4.2               
##  [23] GenomeInfoDbData_1.2.4        bit64_4.0.5                  
##  [25] farver_2.1.0                  vctrs_0.3.8                  
##  [27] generics_0.1.0                xfun_0.23                    
##  [29] BiocFileCache_1.14.0          R6_2.5.0                     
##  [31] doParallel_1.0.16             ggbeeswarm_0.6.0             
##  [33] netSmooth_1.10.0              rsvd_1.0.5                   
##  [35] RJSONIO_1.3-1.4               locfit_1.5-9.4               
##  [37] bitops_1.0-7                  rhdf5filters_1.2.1           
##  [39] cachem_1.0.5                  DelayedArray_0.16.3          
##  [41] assertthat_0.2.1              promises_1.2.0.1             
##  [43] scales_1.1.1                  beeswarm_0.3.1               
##  [45] gtable_0.3.0                  phylobase_0.8.10             
##  [47] beachmat_2.6.4                rlang_0.4.11                 
##  [49] genefilter_1.72.1             splines_4.0.5                
##  [51] lazyeval_0.2.2                BiocManager_1.30.15          
##  [53] yaml_2.2.1                    reshape2_1.4.4               
##  [55] httpuv_1.6.1                  tools_4.0.5                  
##  [57] bookdown_0.22                 gridBase_0.4-7               
##  [59] ellipsis_0.3.2                jquerylib_0.1.4              
##  [61] RColorBrewer_1.1-2            Rcpp_1.0.6                   
##  [63] plyr_1.8.6                    sparseMatrixStats_1.2.1      
##  [65] progress_1.2.2                zlibbioc_1.36.0              
##  [67] purrr_0.3.4                   RCurl_1.98-1.3               
##  [69] prettyunits_1.1.1             viridis_0.6.1                
##  [71] cluster_2.1.1                 tinytex_0.31                 
##  [73] magrittr_2.0.1                data.table_1.14.0            
##  [75] RSpectra_0.16-0               hms_1.1.0                    
##  [77] mime_0.10                     evaluate_0.14                
##  [79] xtable_1.8-4                  XML_3.99-0.6                 
##  [81] gridExtra_2.3                 shape_1.4.6                  
##  [83] compiler_4.0.5                scater_1.18.6                
##  [85] tibble_3.1.2                  RCy3_2.10.2                  
##  [87] crayon_1.4.1                  R.oo_1.24.0                  
##  [89] htmltools_0.5.1.1             entropy_1.3.0                
##  [91] later_1.2.0                   tidyr_1.1.3                  
##  [93] howmany_0.3-1                 DBI_1.1.1                    
##  [95] ExperimentHub_1.16.1          dbplyr_2.1.1                 
##  [97] MASS_7.3-53.1                 rappdirs_0.3.3               
##  [99] Matrix_1.3-2                  ade4_1.7-16                  
## [101] R.methodsS3_1.8.1             igraph_1.2.6                 
## [103] pkgconfig_2.0.3               bigmemory.sri_0.1.3          
## [105] rncl_0.8.4                    registry_0.5-1               
## [107] locfdr_1.1-8                  scuttle_1.0.4                
## [109] xml2_1.3.2                    foreach_1.5.1                
## [111] annotate_1.68.0               vipor_0.4.5                  
## [113] bslib_0.2.5.1                 rngtools_1.5                 
## [115] pkgmaker_0.32.2               XVector_0.30.0               
## [117] stringr_1.4.0                 digest_0.6.27                
## [119] pracma_2.3.3                  graph_1.68.0                 
## [121] softImpute_1.4-1              rmarkdown_2.8                
## [123] edgeR_3.32.1                  DelayedMatrixStats_1.12.3    
## [125] curl_4.3.1                    kernlab_0.9-29               
## [127] shiny_1.6.0                   lifecycle_1.0.0              
## [129] nlme_3.1-152                  jsonlite_1.7.2               
## [131] clusterExperiment_2.10.1      Rhdf5lib_1.12.1              
## [133] BiocNeighbors_1.8.2           viridisLite_0.4.0            
## [135] limma_3.46.0                  fansi_0.4.2                  
## [137] pillar_1.6.1                  lattice_0.20-41              
## [139] fastmap_1.1.0                 httr_1.4.2                   
## [141] survival_3.2-10               interactiveDisplayBase_1.28.0
## [143] glue_1.4.2                    iterators_1.0.13             
## [145] BiocVersion_3.12.0            glmnet_4.1-1                 
## [147] bit_4.0.4                     stringi_1.6.2                
## [149] sass_0.4.0                    HDF5Array_1.18.1             
## [151] blob_1.2.1                    BiocSingular_1.6.0           
## [153] AnnotationHub_2.22.1          memoise_2.0.0                
## [155] dplyr_1.0.6                   irlba_2.3.3                  
## [157] ape_5.5</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="intro.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="lab-2-pathway-level-features.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": false
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/02-BuildPredictor.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
