<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Lab 2: Pathway-level features | Multi-modal Data Integration</title>
  <meta name="description" content="Covers patient classification using multi-modal data integration" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Lab 2: Pathway-level features | Multi-modal Data Integration" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Covers patient classification using multi-modal data integration" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Lab 2: Pathway-level features | Multi-modal Data Integration" />
  
  <meta name="twitter:description" content="Covers patient classification using multi-modal data integration" />
  

<meta name="author" content="Shraddha Pai" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/favicon.ico" type="image/x-icon" />
<link rel="prev" href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"/>

<script src="libs/header-attrs-2.8/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">CBW CAN Multi-modal Data Integration</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Multi-modal Data Integration ( June 2021 )</a></li>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html#class-materials"><i class="fa fa-check"></i>Class materials</a></li>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html#setup-instructions"><i class="fa fa-check"></i>Setup instructions</a></li>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html#basic-programs"><i class="fa fa-check"></i>Basic programs</a>
<ul>
<li class="chapter" data-level="" data-path="welcome.html"><a href="welcome.html#cytoscape-installation"><i class="fa fa-check"></i>Cytoscape Installation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><i class="fa fa-check"></i>Lab 1: Build 3-way patient classifier from 4 data layers</a>
<ul>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#intro-1"><i class="fa fa-check"></i>Intro</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#prepare-data"><i class="fa fa-check"></i>Prepare data</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#make-rules-to-convert-data-into-features"><i class="fa fa-check"></i>Make rules to convert data into features</a>
<ul>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#datalist"><i class="fa fa-check"></i>dataList</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#grouplist"><i class="fa fa-check"></i>groupList</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#define-patient-similarity-for-each-network"><i class="fa fa-check"></i>Define patient similarity for each network</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#build-predictor"><i class="fa fa-check"></i>Build predictor</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#examine-output"><i class="fa fa-check"></i>Examine output</a>
<ul>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#get-model-results"><i class="fa fa-check"></i>Get model results</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#conclusion"><i class="fa fa-check"></i>Conclusion</a></li>
<li class="chapter" data-level="" data-path="lab-1-build-3-way-patient-classifier-from-4-data-layers.html"><a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html#sessioninfo"><i class="fa fa-check"></i>sessionInfo</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html"><i class="fa fa-check"></i>Lab 2: Pathway-level features</a>
<ul>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#setup"><i class="fa fa-check"></i>Setup</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#rules-to-create-features-patient-similarity-networks"><i class="fa fa-check"></i>Rules to create features (patient similarity networks)</a>
<ul>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#rna-pathway-features"><i class="fa fa-check"></i>RNA: Pathway features</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#clinical-single-variables"><i class="fa fa-check"></i>Clinical: Single variables</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#rules-to-create-psns"><i class="fa fa-check"></i>Rules to create PSNs</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#build-predictor-1"><i class="fa fa-check"></i>Build predictor</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#examine-output-1"><i class="fa fa-check"></i>Examine output</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#visualize-top-pathways-in-cytoscape"><i class="fa fa-check"></i>Visualize top pathways in Cytoscape</a></li>
<li class="chapter" data-level="" data-path="lab-2-pathway-level-features.html"><a href="lab-2-pathway-level-features.html#sessioninfo-1"><i class="fa fa-check"></i>sessionInfo</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Multi-modal Data Integration</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="lab-2-pathway-level-features" class="section level1">
<h1>Lab 2: Pathway-level features</h1>
<p>In this example, we will learn how to:</p>
<ul>
<li>Use custom similarity metrics</li>
<li>Group variables into biologically-meaningful units such as pathways, for improved interpretability</li>
</ul>
<p>Here, we will again use breast cancer samples, and discriminate between Luminal A, Luminal B and basal-like tumours. We will limit ourselves to clinical data and gene expression data but will make the following design changes:</p>
<ul>
<li>Clinical variables: A feature (or patient similarity network;PSN) will be defined at the level of selected <em>variables</em> (e.g. age); we will define similarity as <em>normalized difference</em>.</li>
<li>Gene expression: Features will be defined at the level of <strong><em>pathways</em></strong>; i.e. each feature groups only those genes corresponding to the pathway. Similarity is defined as pairwise <em>Pearson correlation</em>.</li>
</ul>
<div class="figure"><span id="fig:unnamed-chunk-27"></span>
<img src="images/Lab2_design.jpg" alt="Lab 2: Predictor design" width="845" />
<p class="caption">
Figure 1: Lab 2: Predictor design
</p>
</div>
<div id="setup" class="section level2">
<h2>Setup</h2>
<p>Let’s fetch the BRCA data using <code>curatedTCGAData</code> again, this time only the gene expression data. Remember, the clinical data we automatically get in the <code>colData()</code> slot returned.</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="lab-2-pathway-level-features.html#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(curatedTCGAData))</span>
<span id="cb69-2"><a href="lab-2-pathway-level-features.html#cb69-2" aria-hidden="true" tabindex="-1"></a>brca <span class="ot">&lt;-</span> <span class="fu">suppressMessages</span>(</span>
<span id="cb69-3"><a href="lab-2-pathway-level-features.html#cb69-3" aria-hidden="true" tabindex="-1"></a>   <span class="fu">curatedTCGAData</span>(</span>
<span id="cb69-4"><a href="lab-2-pathway-level-features.html#cb69-4" aria-hidden="true" tabindex="-1"></a>       <span class="st">&quot;BRCA&quot;</span>,<span class="fu">c</span>(<span class="st">&quot;mRNAArray&quot;</span>),</span>
<span id="cb69-5"><a href="lab-2-pathway-level-features.html#cb69-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">dry.run=</span><span class="cn">FALSE</span>)</span>
<span id="cb69-6"><a href="lab-2-pathway-level-features.html#cb69-6" aria-hidden="true" tabindex="-1"></a>    )</span></code></pre></div>
<p>Let’s look at the data. Notice that we now only have one -omic assay, gene expression measures from microarrays.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="lab-2-pathway-level-features.html#cb70-1" aria-hidden="true" tabindex="-1"></a>brca</span></code></pre></div>
<pre><code>## A MultiAssayExperiment object of 1 listed
##  experiment with a user-defined name and respective class.
##  Containing an ExperimentList class object of length 1:
##  [1] BRCA_mRNAArray-20160128: SummarizedExperiment with 17814 rows and 590 columns
## Functionality:
##  experiments() - obtain the ExperimentList instance
##  colData() - the primary/phenotype DataFrame
##  sampleMap() - the sample coordination DataFrame
##  `$`, `[`, `[[` - extract colData columns, subset, or experiment
##  *Format() - convert into a long or wide DataFrame
##  assays() - convert ExperimentList to a SimpleList of matrices
##  exportClass() - save all data to files</code></pre>
<p>As before, we prepare the data. I highly recommend separating the script that prepares the data from the one running the predictor for improved management, readability and debuggability.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="lab-2-pathway-level-features.html#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;prepare_data.R&quot;</span>)</span>
<span id="cb72-2"><a href="lab-2-pathway-level-features.html#cb72-2" aria-hidden="true" tabindex="-1"></a>brca <span class="ot">&lt;-</span> <span class="fu">prepareDataForCBW</span>(brca, <span class="at">setBinary=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## harmonizing input:
##   removing 44 sampleMap rows with &#39;colname&#39; not in colnames of experiments</code></pre>
</div>
<div id="rules-to-create-features-patient-similarity-networks" class="section level2">
<h2>Rules to create features (patient similarity networks)</h2>
<p>Load the <code>netDx</code> package and initialize the <code>groupList</code> object, where we will store our grouping rules.
Recall that <code>groupList</code> is a list-of-lists, with the top tier containing data layer names, and that the layer names must match <code>names(assays(brca))</code> or whatever your <code>MultiAssayExperiment</code> object is called.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="lab-2-pathway-level-features.html#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressWarnings</span>(<span class="fu">suppressMessages</span>(<span class="fu">require</span>(netDx)))</span>
<span id="cb74-2"><a href="lab-2-pathway-level-features.html#cb74-2" aria-hidden="true" tabindex="-1"></a>groupList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span></code></pre></div>
<div id="rna-pathway-features" class="section level3">
<h3>RNA: Pathway features</h3>
<p>Let’s group genes into <em>pathway-level features</em>, i.e. instead of one PSN for transcriptomic data, we create one PSN for each <em>pathway</em>. So if you had a pathway set with 2,000 curated pathways, <strong>this would generate 2,000 input PSN</strong>.</p>
<p><strong>NOTE:</strong> This design changes model-building time to several hours, so avoid large gene sets (e.g. the full set of ~44,000 Gene Ontology terms, or even ~29,000 GO Biological Process terms). A reasonable start is a compilation of pathways from all curated pathway databases, as in below.
Whicever list you use can be pruned by constraining the min/max number of genes in a set, but the size is something to keep in mind.</p>
<p>There are two ways of providing pathway data to netDx:
1. You can <strong>download</strong> a compilation of pathways from curated databases using the <code>fetchPathwayDefinitions()</code> function in netDx, like so:</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="lab-2-pathway-level-features.html#cb75-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">fetchPathwayDefinitions</span>(<span class="st">&quot;March&quot;</span>,<span class="dv">2021</span>)</span></code></pre></div>
<pre><code>## Fetching http://download.baderlab.org/EM_Genesets/March_01_2021/Human/symbol/Human_AllPathways_March_01_2021_symbol.gmt</code></pre>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="lab-2-pathway-level-features.html#cb77-1" aria-hidden="true" tabindex="-1"></a>x</span></code></pre></div>
<pre><code>##                                                                     BFC2 
## &quot;~/.cache/netDx/20e22fee64ce_Human_AllPathways_March_01_2021_symbol.gmt&quot;</code></pre>
<p>The above pathway set was downloaded from <a href="http://download.baderlab.org/EM_Genesets">download.baderlab.org/EM_Genesets</a>, which is a good resource for routinely-updated curated pathway definitions. Pathways are compiled from Reactome, Panther, NCI, MSigDB, etc., [REF], and data is contained in <a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29">GMT format</a>, a common format to represent gene-sets, such as pathways. You can see from the value of <code>x</code> that the file isn’t downloaded to our working directory, but rather is stored in a BioConductor-specific location for files (file cache), using the <a href="https://www.bioconductor.org/packages/release/bioc/html/BiocFileCache.html">BiocFileCache</a> class. This way if we rerun the script, BioConductor will only download the file again if it has changed since our last download.</p>
<p>Here is an example of the GMT format:</p>
<div class="figure"><span id="fig:unnamed-chunk-33"></span>
<img src="images/GMT_screenshot.png" alt="Lab 2: Predictor design" width="937" />
<p class="caption">
Figure 2: Lab 2: Predictor design
</p>
</div>
<p>We then use <code>readPathways()</code> to read the pathways into a list format to provide the predictor with. So the full call looks like this:</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="lab-2-pathway-level-features.html#cb79-1" aria-hidden="true" tabindex="-1"></a>pathList <span class="ot">&lt;-</span> <span class="fu">readPathways</span>(<span class="fu">fetchPathwayDefinitions</span>(<span class="st">&quot;March&quot;</span>,<span class="dv">2021</span>))</span></code></pre></div>
<pre><code>## ---------------------------------------</code></pre>
<pre><code>## Fetching http://download.baderlab.org/EM_Genesets/March_01_2021/Human/symbol/Human_AllPathways_March_01_2021_symbol.gmt</code></pre>
<pre><code>## File: 20e22fee64ce_Human_AllPathways_March_01_2021_symbol.gmt</code></pre>
<pre><code>## Warning in scan(f, what = &quot;character&quot;, nlines = 1, quiet = TRUE, sep = &quot;\t&quot;):
## EOF within quoted string</code></pre>
<pre><code>## Read 1123 pathways in total, internal list has 1110 entries</code></pre>
<pre><code>##  FILTER: sets with num genes in [10, 200]</code></pre>
<pre><code>##    =&gt; 341 pathways excluded
##    =&gt; 769 left</code></pre>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="lab-2-pathway-level-features.html#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pathList)</span></code></pre></div>
<pre><code>## $STEARATE_BIOSYNTHESIS_I__ANIMALS_
##  [1] &quot;ELOVL1&quot;  &quot;ACOT7&quot;   &quot;ACSL1&quot;   &quot;ACSL5&quot;   &quot;ELOVL6&quot;  &quot;ACSL4&quot;   &quot;ACSL3&quot;  
##  [8] &quot;ACOT2&quot;   &quot;ACOT1&quot;   &quot;ACSBG1&quot;  &quot;ACSBG2&quot;  &quot;SLC27A2&quot; &quot;ACOT4&quot;  
## 
## $SUPERPATHWAY_OF_INOSITOL_PHOSPHATE_COMPOUNDS
##  [1] &quot;PI4K2B&quot;  &quot;MTMR14&quot;  &quot;INPPL1&quot;  &quot;PTEN&quot;    &quot;PIK3CD&quot;  &quot;PIK3C2G&quot; &quot;PIK3CB&quot; 
##  [8] &quot;PIK3C2A&quot; &quot;PIK3CG&quot;  &quot;PIK3C2B&quot; &quot;PIP4P2&quot;  &quot;PIP4P1&quot;  &quot;PLCZ1&quot;   &quot;PPIP5K1&quot;
## [15] &quot;PPIP5K2&quot; &quot;PIP5KL1&quot; &quot;PLCE1&quot;   &quot;PIP4K2A&quot; &quot;PIP4K2B&quot; &quot;PIP4K2C&quot; &quot;SACM1L&quot; 
## [22] &quot;ITPK1&quot;   &quot;IPMK&quot;    &quot;OCRL&quot;    &quot;ITPKB&quot;   &quot;ITPKC&quot;   &quot;MINPP1&quot;  &quot;PLCB3&quot;  
## [29] &quot;PLCB4&quot;   &quot;PIK3CA&quot;  &quot;ITPKA&quot;   &quot;PIK3C3&quot;  &quot;PLCB1&quot;   &quot;PLCB2&quot;   &quot;PI4K2A&quot; 
## [36] &quot;IPPK&quot;    &quot;MTMR3&quot;   &quot;PIK3R4&quot;  &quot;PIK3R3&quot;  &quot;PIK3R2&quot;  &quot;PIK3R1&quot;  &quot;PIK3R6&quot; 
## [43] &quot;PIK3R5&quot;  &quot;INPP5B&quot;  &quot;INPP5A&quot;  &quot;INPP5D&quot;  &quot;PLCG2&quot;   &quot;INPP5J&quot;  &quot;PIP5K1A&quot;
## [50] &quot;INPP5K&quot;  &quot;PIP5K1B&quot; &quot;PIP5K1C&quot; &quot;PLCG1&quot;   &quot;CDIPT&quot;   &quot;IP6K1&quot;   &quot;IP6K3&quot;  
## [57] &quot;IP6K2&quot;   &quot;SYNJ2&quot;   &quot;FIG4&quot;    &quot;PIKFYVE&quot; &quot;SYNJ1&quot;   &quot;PI4KA&quot;   &quot;PLCH1&quot;  
## [64] &quot;PI4KB&quot;   &quot;PLCH2&quot;   &quot;PLCD3&quot;   &quot;PLCD4&quot;   &quot;PLCD1&quot;  
## 
## $PUTRESCINE_DEGRADATION_III
##  [1] &quot;ALDH3A2&quot; &quot;ALDH3B2&quot; &quot;ALDH3A1&quot; &quot;ALDH1B1&quot; &quot;MAOB&quot;    &quot;ALDH2&quot;   &quot;MAOA&quot;   
##  [8] &quot;ALDH3B1&quot; &quot;SAT2&quot;    &quot;SAT1&quot;   
## 
## $TRYPTOPHAN_DEGRADATION_III__EUKARYOTIC_
##  [1] &quot;ACAT1&quot; &quot;HADHB&quot; &quot;GCDH&quot;  &quot;TDO2&quot;  &quot;KYNU&quot;  &quot;HAAO&quot;  &quot;AFMID&quot; &quot;KMO&quot;   &quot;ACAA1&quot;
## [10] &quot;ACAT2&quot; &quot;ACMSD&quot;
## 
## $MEVALONATE_PATHWAY_I
##  [1] &quot;ACAT1&quot;  &quot;IDI1&quot;   &quot;MVK&quot;    &quot;HMGCS1&quot; &quot;IDI2&quot;   &quot;HADHB&quot;  &quot;PMVK&quot;   &quot;MVD&quot;   
##  [9] &quot;HMGCR&quot;  &quot;HMGCS2&quot; &quot;ACAA1&quot;  &quot;ACAT2&quot; 
## 
## $`D-_I_MYO__I_-INOSITOL-5-PHOSPHATE_METABOLISM`
##  [1] &quot;PLCG1&quot;   &quot;PLCB3&quot;   &quot;PLCB4&quot;   &quot;PLCH1&quot;   &quot;PLCH2&quot;   &quot;PLCB1&quot;   &quot;MTMR14&quot; 
##  [8] &quot;PLCD3&quot;   &quot;PLCB2&quot;   &quot;PLCD4&quot;   &quot;PLCD1&quot;   &quot;MTMR3&quot;   &quot;PIP4P2&quot;  &quot;PIP4P1&quot; 
## [15] &quot;PLCZ1&quot;   &quot;PLCG2&quot;   &quot;PLCE1&quot;   &quot;PIP4K2A&quot; &quot;PIP4K2C&quot;</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Alternatively, you can also provide a custom pathway set to netDx by reading in a GMT file, using the <code>readPathways()</code> function. In this example, I’ve downloaded a geneset of <a href="http://www.gsea-msigdb.org/gsea/msigdb/collections.jsp#C7">pathways often dysregulated in cancer</a>, set C6 from MSigDB (REF):</li>
</ol>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="lab-2-pathway-level-features.html#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="do">#####TBA</span></span>
<span id="cb89-2"><a href="lab-2-pathway-level-features.html#cb89-2" aria-hidden="true" tabindex="-1"></a><span class="do">###myGMTfile &lt;- &quot;supporting_files/c6.all.c7.4.symbols.gmt&quot;</span></span>
<span id="cb89-3"><a href="lab-2-pathway-level-features.html#cb89-3" aria-hidden="true" tabindex="-1"></a><span class="do">###file.exists(myGMTfile)</span></span>
<span id="cb89-4"><a href="lab-2-pathway-level-features.html#cb89-4" aria-hidden="true" tabindex="-1"></a><span class="do">####x &lt;- readPathways(myGMTfile)</span></span>
<span id="cb89-5"><a href="lab-2-pathway-level-features.html#cb89-5" aria-hidden="true" tabindex="-1"></a><span class="do">###head(x)</span></span></code></pre></div>
<p>For this tutorial we will limit ourselves to the first set of pathways and use <code>groupList</code> to tell netDx to group transcriptomic data using pathways.</p>
<p><strong>NOTE: The pathway definition file should use the same identifier type as your patient data. For instance, if the genes in your transcriptomic data are represented using <a href="https://www.genenames.org/tools/search/">HGNC symbols</a>, then your pathway definition file must also use HGNC symbols (e.g. <em>ID2S</em>), and not a different type of identifier, such as Ensembl IDs (which look like this: <em>ENSG00000010404</em>).</strong></p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="lab-2-pathway-level-features.html#cb90-1" aria-hidden="true" tabindex="-1"></a>groupList[[<span class="st">&quot;BRCA_mRNAArray-20160128&quot;</span>]] <span class="ot">&lt;-</span> pathList </span></code></pre></div>
</div>
<div id="clinical-single-variables" class="section level3">
<h3>Clinical: Single variables</h3>
<p>Models often include clinical variables such as demographic or disease-related features such as age, sex, or treatment regimen. In this example, we take two variables from the sample metadata, and include create one PSN <em>for each variable</em>.</p>
<p>So here, we map the following:</p>
<ul>
<li><code>patient.age_at_initial_pathologic_diagnosis</code> to <code>age</code> (just shorter)</li>
<li><code>stage</code> to <code>STAGE</code></li>
</ul>
<p>These variables must be present in the <code>colData()</code> slot:</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="lab-2-pathway-level-features.html#cb91-1" aria-hidden="true" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> <span class="fu">colData</span>(brca)</span>
<span id="cb91-2"><a href="lab-2-pathway-level-features.html#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pheno[,<span class="fu">c</span>(<span class="st">&quot;patient.age_at_initial_pathologic_diagnosis&quot;</span>,<span class="st">&quot;STAGE&quot;</span>)])</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 2 columns
##              patient.age_at_initial_pathologic_diagnosis     STAGE
##                                                &lt;integer&gt; &lt;integer&gt;
## TCGA-A1-A0SD                                          59         2
## TCGA-A1-A0SE                                          56         1
## TCGA-A1-A0SH                                          39         2
## TCGA-A1-A0SJ                                          39         3
## TCGA-A1-A0SK                                          54         2
## TCGA-A1-A0SO                                          67         2</code></pre>
<p>We now add the entry into <code>groupList</code>. The entry for <code>clinical</code> is special because netDx will look for corresponding variables in the sample metadata table, <code>colData()</code>, rather than looking for “clinical” within <code>assays(brca)</code>.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="lab-2-pathway-level-features.html#cb93-1" aria-hidden="true" tabindex="-1"></a>groupList[[<span class="st">&quot;clinical&quot;</span>]] <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb93-2"><a href="lab-2-pathway-level-features.html#cb93-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">age=</span><span class="st">&quot;patient.age_at_initial_pathologic_diagnosis&quot;</span>,</span>
<span id="cb93-3"><a href="lab-2-pathway-level-features.html#cb93-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">stage=</span><span class="st">&quot;STAGE&quot;</span></span>
<span id="cb93-4"><a href="lab-2-pathway-level-features.html#cb93-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="rules-to-create-psns" class="section level2">
<h2>Rules to create PSNs</h2>
<p>We now tell netDx what similarity metric to use</p>
<p>This is <code>makeNets()</code> just as we had seen in the previous exercise, which provides <code>netDx</code> with a custom function to generate similarity networks (i.e. features). We previously used the following code to create PSN based on Pearson correlation:</p>
<p><em>(Do not paste this in)</em></p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="lab-2-pathway-level-features.html#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">makePSN_NamedMatrix</span>(..., <span class="at">writeProfiles=</span><span class="cn">TRUE</span>,...)<span class="st">`</span></span></code></pre></div>
<p>We will now make a different call to <code>makePSN_NamedMatrix()</code> but this time, requesting the use of the normalized difference similarity metric.</p>
<p>This is achieved by calling the following code <em>(Do not paste this in, either)</em>:</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="lab-2-pathway-level-features.html#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">makePSN_NamedMatrix</span>(,..., </span>
<span id="cb95-2"><a href="lab-2-pathway-level-features.html#cb95-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">simMetric=</span><span class="st">&quot;custom&quot;</span>, <span class="at">customFunc=</span>normDiff,</span>
<span id="cb95-3"><a href="lab-2-pathway-level-features.html#cb95-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">writeProfiles=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p><code>normDiff</code> is a function provided in the <code>netDx</code> package, but the user may define custom similarity functions in this block of code and pass those to <code>makePSN_NamedMatrix()</code>, using the <code>customFunc</code> parameters; additionally set <code>simMetric</code> to <code>custom</code>.</p>
<p>Other presets provided by netDx include:</p>
<ul>
<li><code>sim.pearscale</code>: Pearson correlation followed by exponential scaling; used with a vector</li>
<li><code>sim.eucscale</code>: Euclidean disance followed by exponential scaling; used with a vector</li>
<li><code>normDiff</code>: Normalized difference; used with a single variable such as age</li>
<li><code>avgNormDiff</code>: Average normalized difference; used with a vector</li>
</ul>
<p>When we’re done, this is what the <code>makeNets</code> function look like.</p>
<p><em>Note: I realize this may be complicated. Future versions of netDx will simplify this syntax.</em></p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="lab-2-pathway-level-features.html#cb96-1" aria-hidden="true" tabindex="-1"></a>makeNets <span class="ot">&lt;-</span> <span class="cf">function</span>(dataList, groupList, netDir,...) {</span>
<span id="cb96-2"><a href="lab-2-pathway-level-features.html#cb96-2" aria-hidden="true" tabindex="-1"></a>    netList <span class="ot">&lt;-</span> <span class="fu">c</span>() </span>
<span id="cb96-3"><a href="lab-2-pathway-level-features.html#cb96-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-4"><a href="lab-2-pathway-level-features.html#cb96-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># make RNA nets (Pearson correlation)</span></span>
<span id="cb96-5"><a href="lab-2-pathway-level-features.html#cb96-5" aria-hidden="true" tabindex="-1"></a>    rna <span class="ot">&lt;-</span> <span class="st">&quot;BRCA_mRNAArray-20160128&quot;</span></span>
<span id="cb96-6"><a href="lab-2-pathway-level-features.html#cb96-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(groupList[[rna]])) { <span class="do">## REMEMBER TO CHECK FOR NULL</span></span>
<span id="cb96-7"><a href="lab-2-pathway-level-features.html#cb96-7" aria-hidden="true" tabindex="-1"></a>        netList <span class="ot">&lt;-</span> <span class="fu">makePSN_NamedMatrix</span>(</span>
<span id="cb96-8"><a href="lab-2-pathway-level-features.html#cb96-8" aria-hidden="true" tabindex="-1"></a>            dataList[[rna]],</span>
<span id="cb96-9"><a href="lab-2-pathway-level-features.html#cb96-9" aria-hidden="true" tabindex="-1"></a>            <span class="fu">rownames</span>(dataList[[rna]]),</span>
<span id="cb96-10"><a href="lab-2-pathway-level-features.html#cb96-10" aria-hidden="true" tabindex="-1"></a>            groupList[[rna]],</span>
<span id="cb96-11"><a href="lab-2-pathway-level-features.html#cb96-11" aria-hidden="true" tabindex="-1"></a>            netDir,</span>
<span id="cb96-12"><a href="lab-2-pathway-level-features.html#cb96-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">verbose=</span><span class="cn">FALSE</span>,</span>
<span id="cb96-13"><a href="lab-2-pathway-level-features.html#cb96-13" aria-hidden="true" tabindex="-1"></a>            <span class="at">writeProfiles=</span><span class="cn">TRUE</span>,         <span class="do">## define Pearson similarity as before</span></span>
<span id="cb96-14"><a href="lab-2-pathway-level-features.html#cb96-14" aria-hidden="true" tabindex="-1"></a>            ...) </span>
<span id="cb96-15"><a href="lab-2-pathway-level-features.html#cb96-15" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb96-16"><a href="lab-2-pathway-level-features.html#cb96-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb96-17"><a href="lab-2-pathway-level-features.html#cb96-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># make clinical nets (normalized difference)</span></span>
<span id="cb96-18"><a href="lab-2-pathway-level-features.html#cb96-18" aria-hidden="true" tabindex="-1"></a>    netList2 <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb96-19"><a href="lab-2-pathway-level-features.html#cb96-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(groupList[[<span class="st">&quot;clinical&quot;</span>]])) {</span>
<span id="cb96-20"><a href="lab-2-pathway-level-features.html#cb96-20" aria-hidden="true" tabindex="-1"></a>    netList2 <span class="ot">&lt;-</span> <span class="fu">makePSN_NamedMatrix</span>(</span>
<span id="cb96-21"><a href="lab-2-pathway-level-features.html#cb96-21" aria-hidden="true" tabindex="-1"></a>        dataList<span class="sc">$</span>clinical, </span>
<span id="cb96-22"><a href="lab-2-pathway-level-features.html#cb96-22" aria-hidden="true" tabindex="-1"></a>        <span class="fu">rownames</span>(dataList<span class="sc">$</span>clinical),</span>
<span id="cb96-23"><a href="lab-2-pathway-level-features.html#cb96-23" aria-hidden="true" tabindex="-1"></a>        groupList[[<span class="st">&quot;clinical&quot;</span>]],netDir,</span>
<span id="cb96-24"><a href="lab-2-pathway-level-features.html#cb96-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">simMetric=</span><span class="st">&quot;custom&quot;</span>,<span class="at">customFunc=</span>normDiff, <span class="do">### Notice simMetric &amp; customFunc</span></span>
<span id="cb96-25"><a href="lab-2-pathway-level-features.html#cb96-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">writeProfiles=</span><span class="cn">FALSE</span>,</span>
<span id="cb96-26"><a href="lab-2-pathway-level-features.html#cb96-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">sparsify=</span><span class="cn">TRUE</span>,</span>
<span id="cb96-27"><a href="lab-2-pathway-level-features.html#cb96-27" aria-hidden="true" tabindex="-1"></a>        <span class="at">verbose=</span><span class="cn">FALSE</span>,</span>
<span id="cb96-28"><a href="lab-2-pathway-level-features.html#cb96-28" aria-hidden="true" tabindex="-1"></a>        ...)</span>
<span id="cb96-29"><a href="lab-2-pathway-level-features.html#cb96-29" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb96-30"><a href="lab-2-pathway-level-features.html#cb96-30" aria-hidden="true" tabindex="-1"></a>    netList <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">unlist</span>(netList),<span class="fu">unlist</span>(netList2))</span>
<span id="cb96-31"><a href="lab-2-pathway-level-features.html#cb96-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(netList)</span>
<span id="cb96-32"><a href="lab-2-pathway-level-features.html#cb96-32" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="build-predictor-1" class="section level2">
<h2>Build predictor</h2>
<p>Finally!
We have:</p>
<ul>
<li>prepared our data,</li>
<li>grouped RNA by pathways (<code>fetchPathwayDefinitions()</code>, <code>readPathways()</code>),</li>
<li>created two PSN using clinical variables (<code>groupList$clinical</code>), and</li>
<li>defined our similarity metrics (<code>makeNets()</code>).</li>
</ul>
<p>Now we build our predictors. For this tutorial, we use two train/test splits. For each split, we assign features a score between zero and two, and call features with score of 1+ “feature-selected.”</p>
<p><strong>Realistic parameters:</strong> When running this with your project, reasonable values are <code>numSplits=10L</code> (10 is a good start, just to see if you get signal, 100+ if firming up for publication), <code>featScoreMax=10L</code>, <code>featSelCutoff=9L</code>.</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="lab-2-pathway-level-features.html#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="do">###t0 &lt;- Sys.time()</span></span>
<span id="cb97-2"><a href="lab-2-pathway-level-features.html#cb97-2" aria-hidden="true" tabindex="-1"></a><span class="do">###set.seed(42) # make results reproducible</span></span>
<span id="cb97-3"><a href="lab-2-pathway-level-features.html#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="do">###outDir &lt;- paste(tempdir(),&quot;pred_output&quot;,sep=getFileSep()) # use absolute path</span></span>
<span id="cb97-4"><a href="lab-2-pathway-level-features.html#cb97-4" aria-hidden="true" tabindex="-1"></a><span class="do">###numSplits &lt;- 2L</span></span>
<span id="cb97-5"><a href="lab-2-pathway-level-features.html#cb97-5" aria-hidden="true" tabindex="-1"></a><span class="do">###model &lt;- suppressMessages(</span></span>
<span id="cb97-6"><a href="lab-2-pathway-level-features.html#cb97-6" aria-hidden="true" tabindex="-1"></a><span class="do">###   buildPredictor(</span></span>
<span id="cb97-7"><a href="lab-2-pathway-level-features.html#cb97-7" aria-hidden="true" tabindex="-1"></a><span class="do">###    dataList=brca,</span></span>
<span id="cb97-8"><a href="lab-2-pathway-level-features.html#cb97-8" aria-hidden="true" tabindex="-1"></a><span class="do">###    groupList=groupList,</span></span>
<span id="cb97-9"><a href="lab-2-pathway-level-features.html#cb97-9" aria-hidden="true" tabindex="-1"></a><span class="do">###    makeNetFunc=makeNets,</span></span>
<span id="cb97-10"><a href="lab-2-pathway-level-features.html#cb97-10" aria-hidden="true" tabindex="-1"></a><span class="do">###    outDir=outDir, </span></span>
<span id="cb97-11"><a href="lab-2-pathway-level-features.html#cb97-11" aria-hidden="true" tabindex="-1"></a><span class="do">###    numSplits=numSplits, </span></span>
<span id="cb97-12"><a href="lab-2-pathway-level-features.html#cb97-12" aria-hidden="true" tabindex="-1"></a><span class="do">###    featScoreMax=2L, </span></span>
<span id="cb97-13"><a href="lab-2-pathway-level-features.html#cb97-13" aria-hidden="true" tabindex="-1"></a><span class="do">###    featSelCutoff=1L,</span></span>
<span id="cb97-14"><a href="lab-2-pathway-level-features.html#cb97-14" aria-hidden="true" tabindex="-1"></a><span class="do">###    numCores=4L</span></span>
<span id="cb97-15"><a href="lab-2-pathway-level-features.html#cb97-15" aria-hidden="true" tabindex="-1"></a><span class="do">###    )</span></span>
<span id="cb97-16"><a href="lab-2-pathway-level-features.html#cb97-16" aria-hidden="true" tabindex="-1"></a><span class="do">###)</span></span>
<span id="cb97-17"><a href="lab-2-pathway-level-features.html#cb97-17" aria-hidden="true" tabindex="-1"></a><span class="do">###t1 &lt;- Sys.time()</span></span>
<span id="cb97-18"><a href="lab-2-pathway-level-features.html#cb97-18" aria-hidden="true" tabindex="-1"></a><span class="do">###print(t1-t0) # time taken</span></span></code></pre></div>
<p>The above settings were chosen to Let’s actually load data generated by reasonable parameters, e.g. <code>numSplits=10L</code>, <code>featScoreMax=10L</code>, <code>featSelCutoff=9L</code>, generated for this tutorial:</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="lab-2-pathway-level-features.html#cb98-1" aria-hidden="true" tabindex="-1"></a>outFile <span class="ot">&lt;-</span> <span class="fu">sprintf</span>(<span class="st">&quot;%s/CBW_Lab2_full.rda&quot;</span>,<span class="fu">tempdir</span>())</span>
<span id="cb98-2"><a href="lab-2-pathway-level-features.html#cb98-2" aria-hidden="true" tabindex="-1"></a>lnames <span class="ot">&lt;-</span> <span class="fu">load</span>(<span class="st">&quot;/home/spai/data/brca_binary_pathways.rda&quot;</span>)</span>
<span id="cb98-3"><a href="lab-2-pathway-level-features.html#cb98-3" aria-hidden="true" tabindex="-1"></a><span class="co">#download.file(&quot;https://github.com/RealPaiLab/CBW_CAN_DataIntegration_2021/raw/master/supporting_files/brca_pathways_full.rda&quot;, </span></span>
<span id="cb98-4"><a href="lab-2-pathway-level-features.html#cb98-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   #destfile=outFile)</span></span>
<span id="cb98-5"><a href="lab-2-pathway-level-features.html#cb98-5" aria-hidden="true" tabindex="-1"></a><span class="co">#lnames &lt;- load(outFile)</span></span></code></pre></div>
<p>We can see the objects in the file using <code>lnames()</code>.</p>
</div>
<div id="examine-output-1" class="section level2">
<h2>Examine output</h2>
<p>As before, we get model results, using <code>getResults()</code> from our helper script, <code>helper.R</code>. We define sel</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="lab-2-pathway-level-features.html#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;helper.R&quot;</span>)</span>
<span id="cb99-2"><a href="lab-2-pathway-level-features.html#cb99-2" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">getResults</span>(brca,model_full,</span>
<span id="cb99-3"><a href="lab-2-pathway-level-features.html#cb99-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">featureSelCutoff=</span>9L,</span>
<span id="cb99-4"><a href="lab-2-pathway-level-features.html#cb99-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">featureSelPct=</span><span class="fl">0.9</span>)</span></code></pre></div>
<pre><code>## Detected 10 splits and 2 classes</code></pre>
<pre><code>## * Plotting performance</code></pre>
<pre><code>## * Compiling feature scores and calling selected features</code></pre>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-44-1.png" width="672" /></p>
<p>Let’s examine our confusion matrix:</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="lab-2-pathway-level-features.html#cb103-1" aria-hidden="true" tabindex="-1"></a>confMat <span class="ot">&lt;-</span> <span class="fu">confusionMatrix</span>(model_full)</span></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/unnamed-chunk-45-1.png" width="672" /></p>
<p>This function returns the confusion matrix for each split, as well as the average shown in the image above:</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="lab-2-pathway-level-features.html#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(confMat)</span></code></pre></div>
<pre><code>##                     Length Class  Mode   
## splitWiseConfMatrix 10     -none- list   
## average              4     table  numeric</code></pre>
</div>
<div id="visualize-top-pathways-in-cytoscape" class="section level2">
<h2>Visualize top pathways in Cytoscape</h2>
<p>We will now visualize top-scoring pathways using the Cytoscape visualization EnrichmentMap. Recall from the XXX module that an EnrichmentMap is a network visualization of related gene-sets, where each node is a pathway and edges connect similar pathways.</p>
<p>Let us say we only want to see pathways that scored in some target range for “most” of the trials; i.e. consistently high-scoring pathways.
We quantify this by asking only for features that score [<code>EMapMinScore</code>,<code>EMapMaxScore</code>] for <code>EMapPctPass</code> fraction of trials.</p>
<p>Here we ask for features scoring 7+ out of 10 for at least 70% of the train/test splits.</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="lab-2-pathway-level-features.html#cb106-1" aria-hidden="true" tabindex="-1"></a>emap <span class="ot">&lt;-</span> <span class="fu">makeInputForEnrichmentMap</span> (</span>
<span id="cb106-2"><a href="lab-2-pathway-level-features.html#cb106-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">model=</span>model_full,</span>
<span id="cb106-3"><a href="lab-2-pathway-level-features.html#cb106-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">results=</span>results,</span>
<span id="cb106-4"><a href="lab-2-pathway-level-features.html#cb106-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">pathwayList=</span>pathList,</span>
<span id="cb106-5"><a href="lab-2-pathway-level-features.html#cb106-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">EMapMinScore=</span>7L, </span>
<span id="cb106-6"><a href="lab-2-pathway-level-features.html#cb106-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">EMapMaxSore=</span>10L,</span>
<span id="cb106-7"><a href="lab-2-pathway-level-features.html#cb106-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">EMapPctPass=</span><span class="fl">0.7</span>,</span>
<span id="cb106-8"><a href="lab-2-pathway-level-features.html#cb106-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">outDir=</span><span class="st">&quot;/home/spai/data&quot;</span></span>
<span id="cb106-9"><a href="lab-2-pathway-level-features.html#cb106-9" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## * Creating input files for EnrichmentMap</code></pre>
<pre><code>## * Writing files for network visualization</code></pre>
<p>This call will return paths to the output files, which you need to now download from AWS to your personal computer:</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="lab-2-pathway-level-features.html#cb109-1" aria-hidden="true" tabindex="-1"></a>emap</span></code></pre></div>
<pre><code>## $GMTfiles
## $GMTfiles$Luminal.A
## [1] &quot;/home/spai/data/Luminal.A.gmt&quot;
## 
## $GMTfiles$other
## [1] &quot;/home/spai/data/other.gmt&quot;
## 
## 
## $NodeStyles
## $NodeStyles$Luminal.A
## [1] &quot;/home/spai/data/Luminal.A_nodeAttrs.txt&quot;
## 
## $NodeStyles$other
## [1] &quot;/home/spai/data/other_nodeAttrs.txt&quot;</code></pre>
<p>Download these now by opening your web browser to: <a href="http://" class="uri">http://</a><your student instance>.</p>
<p>Now let’s use Cytoscape to create an EnrichmentMap.</p>
<ol style="list-style-type: decimal">
<li>Go to Apps &gt; EnrichmentMap.</li>
<li>In the dialog box, click “<strong>+</strong>” on the top-left.</li>
<li>In the right hand panel that opens, select “Analysis type: Generic/gProfiler/Enrichr”</li>
<li>Change “GMT:” to the file you just downloaded: e.g. the file for Luminal.A.</li>
<li>In the bottom panel, click “Show Advanced Options.”</li>
<li>Bottom panel, right: set cutoff = 0.05; metric = Jaccard.</li>
<li>Click “Build.”</li>
</ol>
<p>You should now see a network like this</p>
<p><strong>INSERT NETWORK IMAGE</strong></p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="lab-2-pathway-level-features.html#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="do">###plotEmap(gmtFiles[[1]],nodeAttrFiles[[1]],</span></span>
<span id="cb111-2"><a href="lab-2-pathway-level-features.html#cb111-2" aria-hidden="true" tabindex="-1"></a><span class="do">###         groupClusters=TRUE, hideNodeLabels=TRUE)</span></span></code></pre></div>
</div>
<div id="sessioninfo-1" class="section level2">
<h2>sessionInfo</h2>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="lab-2-pathway-level-features.html#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.0.5 (2021-03-31)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=C                 LC_NUMERIC=C              
##  [3] LC_TIME=C                  LC_COLLATE=C              
##  [5] LC_MONETARY=C              LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] plotrix_3.8-1               ggplot2_3.3.3              
##  [3] Rtsne_0.15                  rhdf5_2.34.0               
##  [5] curatedTCGAData_1.12.1      MultiAssayExperiment_1.16.0
##  [7] SummarizedExperiment_1.20.0 GenomicRanges_1.42.0       
##  [9] GenomeInfoDb_1.26.7         IRanges_2.24.1             
## [11] S4Vectors_0.28.1            MatrixGenerics_1.2.1       
## [13] matrixStats_0.58.0          netDx_1.2.3                
## [15] bigmemory_4.5.36            Biobase_2.50.0             
## [17] BiocGenerics_0.36.1        
## 
## loaded via a namespace (and not attached):
##   [1] utf8_1.2.1                    R.utils_2.10.1               
##   [3] tidyselect_1.1.1              RSQLite_2.2.7                
##   [5] AnnotationDbi_1.52.0          grid_4.0.5                   
##   [7] combinat_0.0-8                BiocParallel_1.24.1          
##   [9] RNeXML_2.4.5                  munsell_0.5.0                
##  [11] codetools_0.2-18              withr_2.4.2                  
##  [13] colorspace_2.0-1              highr_0.9                    
##  [15] knitr_1.33                    uuid_0.1-4                   
##  [17] zinbwave_1.12.0               rstudioapi_0.13              
##  [19] SingleCellExperiment_1.12.0   ROCR_1.0-11                  
##  [21] NMF_0.23.0                    labeling_0.4.2               
##  [23] GenomeInfoDbData_1.2.4        bit64_4.0.5                  
##  [25] farver_2.1.0                  vctrs_0.3.8                  
##  [27] generics_0.1.0                xfun_0.23                    
##  [29] BiocFileCache_1.14.0          R6_2.5.0                     
##  [31] doParallel_1.0.16             ggbeeswarm_0.6.0             
##  [33] netSmooth_1.10.0              rsvd_1.0.5                   
##  [35] RJSONIO_1.3-1.4               locfit_1.5-9.4               
##  [37] bitops_1.0-7                  rhdf5filters_1.2.1           
##  [39] cachem_1.0.5                  DelayedArray_0.16.3          
##  [41] assertthat_0.2.1              promises_1.2.0.1             
##  [43] scales_1.1.1                  beeswarm_0.3.1               
##  [45] gtable_0.3.0                  phylobase_0.8.10             
##  [47] beachmat_2.6.4                rlang_0.4.11                 
##  [49] genefilter_1.72.1             splines_4.0.5                
##  [51] lazyeval_0.2.2                BiocManager_1.30.15          
##  [53] yaml_2.2.1                    reshape2_1.4.4               
##  [55] httpuv_1.6.1                  tools_4.0.5                  
##  [57] bookdown_0.22                 gridBase_0.4-7               
##  [59] ellipsis_0.3.2                jquerylib_0.1.4              
##  [61] RColorBrewer_1.1-2            Rcpp_1.0.6                   
##  [63] plyr_1.8.6                    sparseMatrixStats_1.2.1      
##  [65] progress_1.2.2                zlibbioc_1.36.0              
##  [67] purrr_0.3.4                   RCurl_1.98-1.3               
##  [69] prettyunits_1.1.1             viridis_0.6.1                
##  [71] cluster_2.1.1                 tinytex_0.31                 
##  [73] magrittr_2.0.1                data.table_1.14.0            
##  [75] RSpectra_0.16-0               hms_1.1.0                    
##  [77] mime_0.10                     evaluate_0.14                
##  [79] xtable_1.8-4                  XML_3.99-0.6                 
##  [81] jpeg_0.1-8.1                  gridExtra_2.3                
##  [83] shape_1.4.6                   compiler_4.0.5               
##  [85] scater_1.18.6                 tibble_3.1.2                 
##  [87] RCy3_2.10.2                   crayon_1.4.1                 
##  [89] R.oo_1.24.0                   htmltools_0.5.1.1            
##  [91] entropy_1.3.0                 later_1.2.0                  
##  [93] tidyr_1.1.3                   howmany_0.3-1                
##  [95] DBI_1.1.1                     ExperimentHub_1.16.1         
##  [97] dbplyr_2.1.1                  MASS_7.3-53.1                
##  [99] rappdirs_0.3.3                Matrix_1.3-2                 
## [101] ade4_1.7-16                   R.methodsS3_1.8.1            
## [103] igraph_1.2.6                  pkgconfig_2.0.3              
## [105] bigmemory.sri_0.1.3           rncl_0.8.4                   
## [107] registry_0.5-1                locfdr_1.1-8                 
## [109] scuttle_1.0.4                 xml2_1.3.2                   
## [111] foreach_1.5.1                 annotate_1.68.0              
## [113] vipor_0.4.5                   bslib_0.2.5.1                
## [115] rngtools_1.5                  pkgmaker_0.32.2              
## [117] XVector_0.30.0                stringr_1.4.0                
## [119] digest_0.6.27                 pracma_2.3.3                 
## [121] graph_1.68.0                  softImpute_1.4-1             
## [123] rmarkdown_2.8                 edgeR_3.32.1                 
## [125] DelayedMatrixStats_1.12.3     curl_4.3.1                   
## [127] kernlab_0.9-29                shiny_1.6.0                  
## [129] lifecycle_1.0.0               nlme_3.1-152                 
## [131] jsonlite_1.7.2                clusterExperiment_2.10.1     
## [133] Rhdf5lib_1.12.1               BiocNeighbors_1.8.2          
## [135] viridisLite_0.4.0             limma_3.46.0                 
## [137] fansi_0.4.2                   pillar_1.6.1                 
## [139] lattice_0.20-41               fastmap_1.1.0                
## [141] httr_1.4.2                    survival_3.2-10              
## [143] interactiveDisplayBase_1.28.0 glue_1.4.2                   
## [145] png_0.1-7                     iterators_1.0.13             
## [147] BiocVersion_3.12.0            glmnet_4.1-1                 
## [149] bit_4.0.4                     stringi_1.6.2                
## [151] sass_0.4.0                    HDF5Array_1.18.1             
## [153] blob_1.2.1                    BiocSingular_1.6.0           
## [155] AnnotationHub_2.22.1          memoise_2.0.0                
## [157] dplyr_1.0.6                   irlba_2.3.3                  
## [159] ape_5.5</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="lab-1-build-3-way-patient-classifier-from-4-data-layers.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": false
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/03-PathwayFeatures.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
